// jsonpathly/dist/index.cjs
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);

// src/index.ts
var index_exports = {};
__export(index_exports, {
  JSONPathSyntaxError: () => JSONPathSyntaxError,
  parse: () => parse,
  paths: () => paths,
  query: () => query,
  stringify: () => stringify
});
module.exports = __toCommonJS(index_exports);

// src/parser/errors.ts
var JSONPathSyntaxError = class extends Error {
  constructor(line, charPositionInLine, msg) {
    super(msg);
    this.name = "JSONPathSyntaxError";
    this.line = line;
    this.charPositionInLine = charPositionInLine;
    this.msg = msg;
  }
};

// src/parser/generated/parser.js
function buildSubscriptChain(segments) {
  if (!segments || segments.length === 0) return null;
  let result = null;
  for (let i = segments.length - 1; i >= 0; i--) {
    result = {
      type: "subscript",
      value: segments[i],
      next: result
    };
  }
  return result;
}
function buildBinaryExpr(head, tail, makeNode) {
  return tail.reduce((left, item) => makeNode(left, item), head);
}
var peg$SyntaxError = class extends SyntaxError {
  constructor(message, expected, found, location) {
    super(message);
    this.expected = expected;
    this.found = found;
    this.location = location;
    this.name = "SyntaxError";
  }
  format(sources) {
    let str = "Error: " + this.message;
    if (this.location) {
      let src = null;
      const st = sources.find((s2) => s2.source === this.location.source);
      if (st) {
        src = st.text.split(/\r\n|\n|\r/g);
      }
      const s = this.location.start;
      const offset_s = this.location.source && typeof this.location.source.offset === "function" ? this.location.source.offset(s) : s;
      const loc = this.location.source + ":" + offset_s.line + ":" + offset_s.column;
      if (src) {
        const e = this.location.end;
        const filler = "".padEnd(offset_s.line.toString().length, " ");
        const line = src[s.line - 1];
        const last = s.line === e.line ? e.column : line.length + 1;
        const hatLen = last - s.column || 1;
        str += "\n --> " + loc + "\n" + filler + " |\n" + offset_s.line + " | " + line + "\n" + filler + " | " + "".padEnd(s.column - 1, " ") + "".padEnd(hatLen, "^");
      } else {
        str += "\n at " + loc;
      }
    }
    return str;
  }
  static buildMessage(expected, found) {
    function hex(ch) {
      return ch.codePointAt(0).toString(16).toUpperCase();
    }
    const nonPrintable = Object.prototype.hasOwnProperty.call(RegExp.prototype, "unicode") ? new RegExp("[\\p{C}\\p{Mn}\\p{Mc}]", "gu") : null;
    function unicodeEscape(s) {
      if (nonPrintable) {
        return s.replace(nonPrintable, (ch) => "\\u{" + hex(ch) + "}");
      }
      return s;
    }
    function literalEscape(s) {
      return unicodeEscape(s.replace(/\\/g, "\\\\").replace(/"/g, '\\"').replace(/\0/g, "\\0").replace(/\t/g, "\\t").replace(/\n/g, "\\n").replace(/\r/g, "\\r").replace(/[\x00-\x0F]/g, (ch) => "\\x0" + hex(ch)).replace(/[\x10-\x1F\x7F-\x9F]/g, (ch) => "\\x" + hex(ch)));
    }
    function classEscape(s) {
      return unicodeEscape(s.replace(/\\/g, "\\\\").replace(/\]/g, "\\]").replace(/\^/g, "\\^").replace(/-/g, "\\-").replace(/\0/g, "\\0").replace(/\t/g, "\\t").replace(/\n/g, "\\n").replace(/\r/g, "\\r").replace(/[\x00-\x0F]/g, (ch) => "\\x0" + hex(ch)).replace(/[\x10-\x1F\x7F-\x9F]/g, (ch) => "\\x" + hex(ch)));
    }
    const DESCRIBE_EXPECTATION_FNS = {
      literal(expectation) {
        return '"' + literalEscape(expectation.text) + '"';
      },
      class(expectation) {
        const escapedParts = expectation.parts.map(
          (part) => Array.isArray(part) ? classEscape(part[0]) + "-" + classEscape(part[1]) : classEscape(part)
        );
        return "[" + (expectation.inverted ? "^" : "") + escapedParts.join("") + "]" + (expectation.unicode ? "u" : "");
      },
      any() {
        return "any character";
      },
      end() {
        return "end of input";
      },
      other(expectation) {
        return expectation.description;
      }
    };
    function describeExpectation(expectation) {
      return DESCRIBE_EXPECTATION_FNS[expectation.type](expectation);
    }
    function describeExpected(expected2) {
      const descriptions = expected2.map(describeExpectation);
      descriptions.sort();
      if (descriptions.length > 0) {
        let j = 1;
        for (let i = 1; i < descriptions.length; i++) {
          if (descriptions[i - 1] !== descriptions[i]) {
            descriptions[j] = descriptions[i];
            j++;
          }
        }
        descriptions.length = j;
      }
      switch (descriptions.length) {
        case 1:
          return descriptions[0];
        case 2:
          return descriptions[0] + " or " + descriptions[1];
        default:
          return descriptions.slice(0, -1).join(", ") + ", or " + descriptions[descriptions.length - 1];
      }
    }
    function describeFound(found2) {
      return found2 ? '"' + literalEscape(found2) + '"' : "end of input";
    }
    return "Expected " + describeExpected(expected) + " but " + describeFound(found) + " found.";
  }
};
function peg$parse(input, options) {
  options = options !== void 0 ? options : {};
  const peg$FAILED = {};
  const peg$source = options.grammarSource;
  const peg$startRuleFunctions = {
    jsonpath: peg$parsejsonpath
  };
  let peg$startRuleFunction = peg$parsejsonpath;
  const peg$c0 = "length";
  const peg$c1 = "count";
  const peg$c2 = "match";
  const peg$c3 = "search";
  const peg$c4 = "value";
  const peg$c5 = "$";
  const peg$c6 = "@";
  const peg$c7 = "..";
  const peg$c8 = ".";
  const peg$c9 = "*";
  const peg$c10 = "&&";
  const peg$c11 = "||";
  const peg$c12 = "!";
  const peg$c13 = "==";
  const peg$c14 = "!=";
  const peg$c15 = ">=";
  const peg$c16 = ">";
  const peg$c17 = "<=";
  const peg$c18 = "<";
  const peg$c19 = "=~";
  const peg$c20 = "in";
  const peg$c21 = "nin";
  const peg$c22 = "subsetof";
  const peg$c23 = "anyof";
  const peg$c24 = "noneof";
  const peg$c25 = "sizeof";
  const peg$c26 = "size";
  const peg$c27 = "empty";
  const peg$c28 = "true";
  const peg$c29 = "false";
  const peg$c30 = "null";
  const peg$c31 = "{";
  const peg$c32 = "}";
  const peg$c33 = "[";
  const peg$c34 = "]";
  const peg$c35 = ":";
  const peg$c36 = ",";
  const peg$c37 = "(";
  const peg$c38 = ")";
  const peg$c39 = "?";
  const peg$c40 = "/";
  const peg$c41 = "\\";
  const peg$c42 = "-";
  const peg$c43 = '"';
  const peg$c44 = "'";
  const peg$c45 = "b";
  const peg$c46 = "f";
  const peg$c47 = "n";
  const peg$c48 = "r";
  const peg$c49 = "t";
  const peg$c50 = "u";
  const peg$r0 = /^[+\-]/;
  const peg$r1 = /^[*\/]/;
  const peg$r2 = /^[^\/\\]/;
  const peg$r3 = /^[gimsuy]/;
  const peg$r4 = /^[a-zA-Z_]/;
  const peg$r5 = /^[a-zA-Z0-9_]/;
  const peg$r6 = /^[0-9@$\-]/;
  const peg$r7 = /^[\x80-\uFFFF_]/;
  const peg$r8 = /^[0-9]/;
  const peg$r9 = /^[eE]/;
  const peg$r10 = /^["\\]/;
  const peg$r11 = /^['\\]/;
  const peg$r12 = /^[0-9a-fA-F]/;
  const peg$r13 = /^[ \t\n\r]/;
  const peg$e0 = peg$literalExpectation("length", true);
  const peg$e1 = peg$literalExpectation("count", true);
  const peg$e2 = peg$literalExpectation("match", true);
  const peg$e3 = peg$literalExpectation("search", true);
  const peg$e4 = peg$literalExpectation("value", true);
  const peg$e5 = peg$classExpectation(["+", "-"], false, false, false);
  const peg$e6 = peg$classExpectation(["*", "/"], false, false, false);
  const peg$e7 = peg$otherExpectation("string");
  const peg$e8 = peg$literalExpectation("$", false);
  const peg$e9 = peg$literalExpectation("@", false);
  const peg$e10 = peg$literalExpectation("..", false);
  const peg$e11 = peg$literalExpectation(".", false);
  const peg$e12 = peg$literalExpectation("*", false);
  const peg$e13 = peg$literalExpectation("&&", false);
  const peg$e14 = peg$literalExpectation("||", false);
  const peg$e15 = peg$literalExpectation("!", false);
  const peg$e16 = peg$literalExpectation("==", false);
  const peg$e17 = peg$literalExpectation("!=", false);
  const peg$e18 = peg$literalExpectation(">=", false);
  const peg$e19 = peg$literalExpectation(">", false);
  const peg$e20 = peg$literalExpectation("<=", false);
  const peg$e21 = peg$literalExpectation("<", false);
  const peg$e22 = peg$literalExpectation("=~", false);
  const peg$e23 = peg$literalExpectation("in", false);
  const peg$e24 = peg$literalExpectation("nin", false);
  const peg$e25 = peg$literalExpectation("subsetof", false);
  const peg$e26 = peg$literalExpectation("anyof", false);
  const peg$e27 = peg$literalExpectation("noneof", false);
  const peg$e28 = peg$literalExpectation("sizeof", false);
  const peg$e29 = peg$literalExpectation("size", false);
  const peg$e30 = peg$literalExpectation("empty", false);
  const peg$e31 = peg$literalExpectation("true", false);
  const peg$e32 = peg$literalExpectation("false", false);
  const peg$e33 = peg$literalExpectation("null", false);
  const peg$e34 = peg$literalExpectation("{", false);
  const peg$e35 = peg$literalExpectation("}", false);
  const peg$e36 = peg$literalExpectation("[", false);
  const peg$e37 = peg$literalExpectation("]", false);
  const peg$e38 = peg$literalExpectation(":", false);
  const peg$e39 = peg$literalExpectation(",", false);
  const peg$e40 = peg$literalExpectation("(", false);
  const peg$e41 = peg$literalExpectation(")", false);
  const peg$e42 = peg$literalExpectation("?", false);
  const peg$e43 = peg$otherExpectation("regex");
  const peg$e44 = peg$literalExpectation("/", false);
  const peg$e45 = peg$classExpectation(["/", "\\"], true, false, false);
  const peg$e46 = peg$literalExpectation("\\", false);
  const peg$e47 = peg$anyExpectation();
  const peg$e48 = peg$classExpectation(["g", "i", "m", "s", "u", "y"], false, false, false);
  const peg$e49 = peg$classExpectation([["a", "z"], ["A", "Z"], "_"], false, false, false);
  const peg$e50 = peg$classExpectation([["a", "z"], ["A", "Z"], ["0", "9"], "_"], false, false, false);
  const peg$e51 = peg$literalExpectation("-", false);
  const peg$e52 = peg$classExpectation([["0", "9"], "@", "$", "-"], false, false, false);
  const peg$e53 = peg$classExpectation([["\x80", "\uFFFF"], "_"], false, false, false);
  const peg$e54 = peg$otherExpectation("number");
  const peg$e55 = peg$classExpectation([["0", "9"]], false, false, false);
  const peg$e56 = peg$classExpectation(["e", "E"], false, false, false);
  const peg$e57 = peg$literalExpectation('"', false);
  const peg$e58 = peg$literalExpectation("'", false);
  const peg$e59 = peg$classExpectation(['"', "\\"], false, false, false);
  const peg$e60 = peg$classExpectation(["'", "\\"], false, false, false);
  const peg$e61 = peg$literalExpectation("b", false);
  const peg$e62 = peg$literalExpectation("f", false);
  const peg$e63 = peg$literalExpectation("n", false);
  const peg$e64 = peg$literalExpectation("r", false);
  const peg$e65 = peg$literalExpectation("t", false);
  const peg$e66 = peg$literalExpectation("u", false);
  const peg$e67 = peg$classExpectation([["0", "9"], ["a", "f"], ["A", "F"]], false, false, false);
  const peg$e68 = peg$classExpectation([" ", "	", "\n", "\r"], false, false, false);
  function peg$f0(segments) {
    return {
      type: "root",
      next: buildSubscriptChain(segments)
    };
  }
  function peg$f1(content) {
    return { type: "dot", value: content };
  }
  function peg$f2(bracket) {
    return bracket;
  }
  function peg$f3(content) {
    return { type: "dotdot", value: content };
  }
  function peg$f4() {
    return { type: "wildcard" };
  }
  function peg$f5(n) {
    return { type: "numericLiteral", value: n };
  }
  function peg$f6(id) {
    return { type: "identifier", value: id };
  }
  function peg$f7() {
    return { type: "wildcard" };
  }
  function peg$f8(id) {
    return { type: "identifier", value: id };
  }
  function peg$f9(bracket) {
    return bracket;
  }
  function peg$f10(content) {
    return content;
  }
  function peg$f11() {
    return { type: "bracketExpression", value: { type: "wildcard" } };
  }
  function peg$f12(n) {
    return { type: "bracketMember", value: { type: "numericLiteral", value: n } };
  }
  function peg$f13(s) {
    return { type: "bracketMember", value: { type: "stringLiteral", value: s } };
  }
  function peg$f14(id) {
    return { type: "bracketMember", value: { type: "identifier", value: id } };
  }
  function peg$f15(expr) {
    return { type: "bracketExpression", value: { type: "filterExpression", value: expr } };
  }
  function peg$f16(expr) {
    return { type: "bracketExpression", value: { type: "filterExpression", value: expr } };
  }
  function peg$f17(head, tail) {
    return buildBinaryExpr(head, tail, (left, item) => ({
      type: "logicalExpression",
      operator: "or",
      left,
      right: item[3]
    }));
  }
  function peg$f18(head, tail) {
    return buildBinaryExpr(head, tail, (left, item) => ({
      type: "logicalExpression",
      operator: "and",
      left,
      right: item[3]
    }));
  }
  function peg$f19(expr) {
    return { type: "notExpression", value: expr };
  }
  function peg$f20(expr) {
    return { type: "notExpression", value: expr };
  }
  function peg$f21(left, op, right) {
    return { type: "comparator", operator: op, left, right };
  }
  function peg$f22(left, regex) {
    return { type: "comparator", operator: "reg", left, right: regex };
  }
  function peg$f23(left) {
    return { type: "comparator", operator: "empty", left, right: null };
  }
  function peg$f24() {
    return "eq";
  }
  function peg$f25() {
    return "ne";
  }
  function peg$f26() {
    return "le";
  }
  function peg$f27() {
    return "lt";
  }
  function peg$f28() {
    return "ge";
  }
  function peg$f29() {
    return "gt";
  }
  function peg$f30() {
    return "in";
  }
  function peg$f31() {
    return "nin";
  }
  function peg$f32() {
    return "subsetof";
  }
  function peg$f33() {
    return "anyof";
  }
  function peg$f34() {
    return "noneof";
  }
  function peg$f35() {
    return "sizeof";
  }
  function peg$f36() {
    return "size";
  }
  function peg$f37(expr) {
    return { type: "groupExpression", value: expr };
  }
  function peg$f38() {
    return { type: "value", value: true, subtype: "boolean" };
  }
  function peg$f39() {
    return { type: "value", value: false, subtype: "boolean" };
  }
  function peg$f40() {
    return { type: "value", value: null, subtype: "null" };
  }
  function peg$f41(name, args) {
    return { type: "functionCall", name, args: args || [] };
  }
  function peg$f42() {
    return "length";
  }
  function peg$f43() {
    return "count";
  }
  function peg$f44() {
    return "match";
  }
  function peg$f45() {
    return "search";
  }
  function peg$f46() {
    return "value";
  }
  function peg$f47(head, tail) {
    return [head, ...tail.map((t) => t[3])];
  }
  function peg$f48(head, tail) {
    if (tail.length === 0) return head;
    return buildBinaryExpr(head, tail, (left, item) => ({
      type: "operation",
      operator: item[1] === "+" ? "plus" : "minus",
      left,
      right: item[3]
    }));
  }
  function peg$f49(head, tail) {
    if (tail.length === 0) return head;
    return buildBinaryExpr(head, tail, (left, item) => ({
      type: "operation",
      operator: item[1] === "*" ? "multi" : "div",
      left,
      right: item[3]
    }));
  }
  function peg$f50(arg) {
    return { type: "groupOperation", value: arg };
  }
  function peg$f51(segments) {
    return {
      type: "root",
      next: buildSubscriptChain(segments)
    };
  }
  function peg$f52(segments) {
    return {
      type: "current",
      next: buildSubscriptChain(segments)
    };
  }
  function peg$f53(start, end, step) {
    return {
      type: "bracketExpression",
      value: {
        type: "slices",
        start: start !== null ? start : null,
        end: end !== null ? end : null,
        step: step ? step[1] !== null ? step[1] : null : null
      }
    };
  }
  function peg$f54(head, tail) {
    const values = [head, ...tail.map((t) => t[3])];
    return { type: "bracketExpression", value: { type: "unions", values } };
  }
  function peg$f55() {
    return { type: "wildcard" };
  }
  function peg$f56(n) {
    return { type: "numericLiteral", value: n };
  }
  function peg$f57(s) {
    return { type: "stringLiteral", value: s };
  }
  function peg$f58(id) {
    return { type: "identifier", value: id };
  }
  function peg$f59(expr, opts) {
    return { type: "value", subtype: "regex", value: expr, opts: opts || "" };
  }
  function peg$f60(s) {
    return { type: "value", subtype: "string", value: s };
  }
  function peg$f61(n) {
    return { type: "value", subtype: "number", value: n };
  }
  function peg$f62() {
    return { type: "value", value: true, subtype: "boolean" };
  }
  function peg$f63() {
    return { type: "value", value: false, subtype: "boolean" };
  }
  function peg$f64() {
    return { type: "value", value: null, subtype: "null" };
  }
  function peg$f65(s) {
    return s;
  }
  function peg$f66(pairs) {
    const obj = {};
    if (pairs) {
      for (const [key, val] of pairs) {
        obj[key] = val;
      }
    }
    return { type: "value", subtype: "object", value: obj };
  }
  function peg$f67(head, tail) {
    return [head, ...tail.map((t) => t[3])];
  }
  function peg$f68(key, val) {
    return [key, val.value];
  }
  function peg$f69(items) {
    return { type: "value", subtype: "array", value: items || [] };
  }
  function peg$f70(head, tail) {
    return [head.value, ...tail.map((t) => t[3].value)];
  }
  function peg$f71(chars) {
    return chars;
  }
  function peg$f72(chars) {
    return chars;
  }
  function peg$f73() {
    return "true";
  }
  function peg$f74() {
    return "false";
  }
  function peg$f75() {
    return "null";
  }
  function peg$f76(chars) {
    return "/" + chars.map((c) => Array.isArray(c) ? c.join("") : c).join("") + "/";
  }
  function peg$f77(opts) {
    return opts.join("");
  }
  function peg$f78(first, rest) {
    return first + rest.join("");
  }
  function peg$f79() {
    return "-";
  }
  function peg$f80(chars) {
    return chars.join("");
  }
  function peg$f81(frac, exp) {
    const num = parseFloat(text());
    if (!frac && !exp && (num < -9007199254740991 || num > 9007199254740991)) {
      error("Integer out of I-JSON range [-(2^53)+1, (2^53)-1]");
    }
    return num;
  }
  function peg$f82(chars) {
    return chars.join("");
  }
  function peg$f83(chars) {
    return chars.join("");
  }
  function peg$f84(char) {
    return char;
  }
  function peg$f85(seq) {
    return seq;
  }
  function peg$f86(char) {
    return char;
  }
  function peg$f87(seq) {
    return seq;
  }
  function peg$f88() {
    return '"';
  }
  function peg$f89() {
    return "'";
  }
  function peg$f90() {
    return "\\";
  }
  function peg$f91() {
    return "/";
  }
  function peg$f92() {
    return "\b";
  }
  function peg$f93() {
    return "\f";
  }
  function peg$f94() {
    return "\n";
  }
  function peg$f95() {
    return "\r";
  }
  function peg$f96() {
    return "	";
  }
  function peg$f97(digits) {
    return String.fromCharCode(parseInt(digits, 16));
  }
  let peg$currPos = options.peg$currPos | 0;
  let peg$savedPos = peg$currPos;
  const peg$posDetailsCache = [{ line: 1, column: 1 }];
  let peg$maxFailPos = peg$currPos;
  let peg$maxFailExpected = options.peg$maxFailExpected || [];
  let peg$silentFails = options.peg$silentFails | 0;
  let peg$result;
  if (options.startRule) {
    if (!(options.startRule in peg$startRuleFunctions)) {
      throw new Error(`Can't start parsing from rule "` + options.startRule + '".');
    }
    peg$startRuleFunction = peg$startRuleFunctions[options.startRule];
  }
  function text() {
    return input.substring(peg$savedPos, peg$currPos);
  }
  function offset() {
    return peg$savedPos;
  }
  function range() {
    return {
      source: peg$source,
      start: peg$savedPos,
      end: peg$currPos
    };
  }
  function location() {
    return peg$computeLocation(peg$savedPos, peg$currPos);
  }
  function expected(description, location2) {
    location2 = location2 !== void 0 ? location2 : peg$computeLocation(peg$savedPos, peg$currPos);
    throw peg$buildStructuredError(
      [peg$otherExpectation(description)],
      input.substring(peg$savedPos, peg$currPos),
      location2
    );
  }
  function error(message, location2) {
    location2 = location2 !== void 0 ? location2 : peg$computeLocation(peg$savedPos, peg$currPos);
    throw peg$buildSimpleError(message, location2);
  }
  function peg$getUnicode(pos = peg$currPos) {
    const cp = input.codePointAt(pos);
    if (cp === void 0) {
      return "";
    }
    return String.fromCodePoint(cp);
  }
  function peg$literalExpectation(text2, ignoreCase) {
    return { type: "literal", text: text2, ignoreCase };
  }
  function peg$classExpectation(parts, inverted, ignoreCase, unicode) {
    return { type: "class", parts, inverted, ignoreCase, unicode };
  }
  function peg$anyExpectation() {
    return { type: "any" };
  }
  function peg$endExpectation() {
    return { type: "end" };
  }
  function peg$otherExpectation(description) {
    return { type: "other", description };
  }
  function peg$computePosDetails(pos) {
    let details = peg$posDetailsCache[pos];
    let p;
    if (details) {
      return details;
    } else {
      if (pos >= peg$posDetailsCache.length) {
        p = peg$posDetailsCache.length - 1;
      } else {
        p = pos;
        while (!peg$posDetailsCache[--p]) {
        }
      }
      details = peg$posDetailsCache[p];
      details = {
        line: details.line,
        column: details.column
      };
      while (p < pos) {
        if (input.charCodeAt(p) === 10) {
          details.line++;
          details.column = 1;
        } else {
          details.column++;
        }
        p++;
      }
      peg$posDetailsCache[pos] = details;
      return details;
    }
  }
  function peg$computeLocation(startPos, endPos, offset2) {
    const startPosDetails = peg$computePosDetails(startPos);
    const endPosDetails = peg$computePosDetails(endPos);
    const res = {
      source: peg$source,
      start: {
        offset: startPos,
        line: startPosDetails.line,
        column: startPosDetails.column
      },
      end: {
        offset: endPos,
        line: endPosDetails.line,
        column: endPosDetails.column
      }
    };
    if (offset2 && peg$source && typeof peg$source.offset === "function") {
      res.start = peg$source.offset(res.start);
      res.end = peg$source.offset(res.end);
    }
    return res;
  }
  function peg$fail(expected2) {
    if (peg$currPos < peg$maxFailPos) {
      return;
    }
    if (peg$currPos > peg$maxFailPos) {
      peg$maxFailPos = peg$currPos;
      peg$maxFailExpected = [];
    }
    peg$maxFailExpected.push(expected2);
  }
  function peg$buildSimpleError(message, location2) {
    return new peg$SyntaxError(message, null, null, location2);
  }
  function peg$buildStructuredError(expected2, found, location2) {
    return new peg$SyntaxError(
      peg$SyntaxError.buildMessage(expected2, found),
      expected2,
      found,
      location2
    );
  }
  function peg$parsejsonpath() {
    let s0, s1, s2, s3, s4, s5;
    s0 = peg$currPos;
    s1 = peg$parse_();
    s2 = peg$parseROOT_VALUE();
    if (s2 !== peg$FAILED) {
      s3 = [];
      s4 = peg$parsesegment();
      while (s4 !== peg$FAILED) {
        s3.push(s4);
        s4 = peg$parsesegment();
      }
      s4 = peg$parse_();
      s5 = peg$parseEOF();
      if (s5 !== peg$FAILED) {
        peg$savedPos = s0;
        s0 = peg$f0(s3);
      } else {
        peg$currPos = s0;
        s0 = peg$FAILED;
      }
    } else {
      peg$currPos = s0;
      s0 = peg$FAILED;
    }
    return s0;
  }
  function peg$parsesegment() {
    let s0;
    s0 = peg$parsedescendantSegment();
    if (s0 === peg$FAILED) {
      s0 = peg$parsechildSegment();
    }
    return s0;
  }
  function peg$parsechildSegment() {
    let s0, s1, s2;
    s0 = peg$currPos;
    s1 = peg$parseDOT();
    if (s1 !== peg$FAILED) {
      s2 = peg$parsedotContent();
      if (s2 !== peg$FAILED) {
        peg$savedPos = s0;
        s0 = peg$f1(s2);
      } else {
        peg$currPos = s0;
        s0 = peg$FAILED;
      }
    } else {
      peg$currPos = s0;
      s0 = peg$FAILED;
    }
    if (s0 === peg$FAILED) {
      s0 = peg$currPos;
      s1 = peg$parsebracket();
      if (s1 !== peg$FAILED) {
        peg$savedPos = s0;
        s1 = peg$f2(s1);
      }
      s0 = s1;
    }
    return s0;
  }
  function peg$parsedescendantSegment() {
    let s0, s1, s2;
    s0 = peg$currPos;
    s1 = peg$parseDOTDOT();
    if (s1 !== peg$FAILED) {
      s2 = peg$parsedotdotContent();
      if (s2 !== peg$FAILED) {
        peg$savedPos = s0;
        s0 = peg$f3(s2);
      } else {
        peg$currPos = s0;
        s0 = peg$FAILED;
      }
    } else {
      peg$currPos = s0;
      s0 = peg$FAILED;
    }
    return s0;
  }
  function peg$parsedotContent() {
    let s0, s1;
    s0 = peg$currPos;
    s1 = peg$parseSTAR();
    if (s1 !== peg$FAILED) {
      peg$savedPos = s0;
      s1 = peg$f4();
    }
    s0 = s1;
    if (s0 === peg$FAILED) {
      s0 = peg$currPos;
      s1 = peg$parseNUMBER();
      if (s1 !== peg$FAILED) {
        peg$savedPos = s0;
        s1 = peg$f5(s1);
      }
      s0 = s1;
      if (s0 === peg$FAILED) {
        s0 = peg$currPos;
        s1 = peg$parseidentifier();
        if (s1 !== peg$FAILED) {
          peg$savedPos = s0;
          s1 = peg$f6(s1);
        }
        s0 = s1;
      }
    }
    return s0;
  }
  function peg$parsedotdotContent() {
    let s0, s1;
    s0 = peg$currPos;
    s1 = peg$parseSTAR();
    if (s1 !== peg$FAILED) {
      peg$savedPos = s0;
      s1 = peg$f7();
    }
    s0 = s1;
    if (s0 === peg$FAILED) {
      s0 = peg$currPos;
      s1 = peg$parseidentifier();
      if (s1 !== peg$FAILED) {
        peg$savedPos = s0;
        s1 = peg$f8(s1);
      }
      s0 = s1;
      if (s0 === peg$FAILED) {
        s0 = peg$currPos;
        s1 = peg$parsebracket();
        if (s1 !== peg$FAILED) {
          peg$savedPos = s0;
          s1 = peg$f9(s1);
        }
        s0 = s1;
      }
    }
    return s0;
  }
  function peg$parsebracket() {
    let s0, s1, s2, s3, s4, s5;
    s0 = peg$currPos;
    s1 = peg$parseBRACKET_LEFT();
    if (s1 !== peg$FAILED) {
      s2 = peg$parse_();
      s3 = peg$parsebracketContent();
      if (s3 !== peg$FAILED) {
        s4 = peg$parse_();
        s5 = peg$parseBRACKET_RIGHT();
        if (s5 !== peg$FAILED) {
          peg$savedPos = s0;
          s0 = peg$f10(s3);
        } else {
          peg$currPos = s0;
          s0 = peg$FAILED;
        }
      } else {
        peg$currPos = s0;
        s0 = peg$FAILED;
      }
    } else {
      peg$currPos = s0;
      s0 = peg$FAILED;
    }
    return s0;
  }
  function peg$parsebracketContent() {
    let s0, s1;
    s0 = peg$parsefilterExpression();
    if (s0 === peg$FAILED) {
      s0 = peg$parseslices();
      if (s0 === peg$FAILED) {
        s0 = peg$parseselectorList();
        if (s0 === peg$FAILED) {
          s0 = peg$currPos;
          s1 = peg$parseSTAR();
          if (s1 !== peg$FAILED) {
            peg$savedPos = s0;
            s1 = peg$f11();
          }
          s0 = s1;
          if (s0 === peg$FAILED) {
            s0 = peg$currPos;
            s1 = peg$parseNUMBER();
            if (s1 !== peg$FAILED) {
              peg$savedPos = s0;
              s1 = peg$f12(s1);
            }
            s0 = s1;
            if (s0 === peg$FAILED) {
              s0 = peg$currPos;
              s1 = peg$parseSTRING();
              if (s1 !== peg$FAILED) {
                peg$savedPos = s0;
                s1 = peg$f13(s1);
              }
              s0 = s1;
              if (s0 === peg$FAILED) {
                s0 = peg$currPos;
                s1 = peg$parseidentifier();
                if (s1 !== peg$FAILED) {
                  peg$savedPos = s0;
                  s1 = peg$f14(s1);
                }
                s0 = s1;
              }
            }
          }
        }
      }
    }
    return s0;
  }
  function peg$parsefilterExpression() {
    let s0, s1, s2, s3, s4, s5, s6, s7;
    s0 = peg$currPos;
    s1 = peg$parseQUESTION();
    if (s1 !== peg$FAILED) {
      s2 = peg$parse_();
      s3 = peg$parsePAREN_LEFT();
      if (s3 !== peg$FAILED) {
        s4 = peg$parse_();
        s5 = peg$parseorExpression();
        if (s5 !== peg$FAILED) {
          s6 = peg$parse_();
          s7 = peg$parsePAREN_RIGHT();
          if (s7 !== peg$FAILED) {
            peg$savedPos = s0;
            s0 = peg$f15(s5);
          } else {
            peg$currPos = s0;
            s0 = peg$FAILED;
          }
        } else {
          peg$currPos = s0;
          s0 = peg$FAILED;
        }
      } else {
        peg$currPos = s0;
        s0 = peg$FAILED;
      }
    } else {
      peg$currPos = s0;
      s0 = peg$FAILED;
    }
    if (s0 === peg$FAILED) {
      s0 = peg$currPos;
      s1 = peg$parseQUESTION();
      if (s1 !== peg$FAILED) {
        s2 = peg$parse_();
        s3 = peg$parseorExpression();
        if (s3 !== peg$FAILED) {
          peg$savedPos = s0;
          s0 = peg$f16(s3);
        } else {
          peg$currPos = s0;
          s0 = peg$FAILED;
        }
      } else {
        peg$currPos = s0;
        s0 = peg$FAILED;
      }
    }
    return s0;
  }
  function peg$parseorExpression() {
    let s0, s1, s2, s3, s4, s5, s6, s7;
    s0 = peg$currPos;
    s1 = peg$parseandExpression();
    if (s1 !== peg$FAILED) {
      s2 = [];
      s3 = peg$currPos;
      s4 = peg$parse_();
      s5 = peg$parseOR();
      if (s5 !== peg$FAILED) {
        s6 = peg$parse_();
        s7 = peg$parseandExpression();
        if (s7 !== peg$FAILED) {
          s4 = [s4, s5, s6, s7];
          s3 = s4;
        } else {
          peg$currPos = s3;
          s3 = peg$FAILED;
        }
      } else {
        peg$currPos = s3;
        s3 = peg$FAILED;
      }
      while (s3 !== peg$FAILED) {
        s2.push(s3);
        s3 = peg$currPos;
        s4 = peg$parse_();
        s5 = peg$parseOR();
        if (s5 !== peg$FAILED) {
          s6 = peg$parse_();
          s7 = peg$parseandExpression();
          if (s7 !== peg$FAILED) {
            s4 = [s4, s5, s6, s7];
            s3 = s4;
          } else {
            peg$currPos = s3;
            s3 = peg$FAILED;
          }
        } else {
          peg$currPos = s3;
          s3 = peg$FAILED;
        }
      }
      peg$savedPos = s0;
      s0 = peg$f17(s1, s2);
    } else {
      peg$currPos = s0;
      s0 = peg$FAILED;
    }
    return s0;
  }
  function peg$parseandExpression() {
    let s0, s1, s2, s3, s4, s5, s6, s7;
    s0 = peg$currPos;
    s1 = peg$parsenotExpression();
    if (s1 !== peg$FAILED) {
      s2 = [];
      s3 = peg$currPos;
      s4 = peg$parse_();
      s5 = peg$parseAND();
      if (s5 !== peg$FAILED) {
        s6 = peg$parse_();
        s7 = peg$parsenotExpression();
        if (s7 !== peg$FAILED) {
          s4 = [s4, s5, s6, s7];
          s3 = s4;
        } else {
          peg$currPos = s3;
          s3 = peg$FAILED;
        }
      } else {
        peg$currPos = s3;
        s3 = peg$FAILED;
      }
      while (s3 !== peg$FAILED) {
        s2.push(s3);
        s3 = peg$currPos;
        s4 = peg$parse_();
        s5 = peg$parseAND();
        if (s5 !== peg$FAILED) {
          s6 = peg$parse_();
          s7 = peg$parsenotExpression();
          if (s7 !== peg$FAILED) {
            s4 = [s4, s5, s6, s7];
            s3 = s4;
          } else {
            peg$currPos = s3;
            s3 = peg$FAILED;
          }
        } else {
          peg$currPos = s3;
          s3 = peg$FAILED;
        }
      }
      peg$savedPos = s0;
      s0 = peg$f18(s1, s2);
    } else {
      peg$currPos = s0;
      s0 = peg$FAILED;
    }
    return s0;
  }
  function peg$parsenotExpression() {
    let s0, s1, s2, s3, s4, s5, s6, s7;
    s0 = peg$currPos;
    s1 = peg$parseNOT();
    if (s1 !== peg$FAILED) {
      s2 = peg$parse_();
      s3 = peg$parsePAREN_LEFT();
      if (s3 !== peg$FAILED) {
        s4 = peg$parse_();
        s5 = peg$parseorExpression();
        if (s5 !== peg$FAILED) {
          s6 = peg$parse_();
          s7 = peg$parsePAREN_RIGHT();
          if (s7 !== peg$FAILED) {
            peg$savedPos = s0;
            s0 = peg$f19(s5);
          } else {
            peg$currPos = s0;
            s0 = peg$FAILED;
          }
        } else {
          peg$currPos = s0;
          s0 = peg$FAILED;
        }
      } else {
        peg$currPos = s0;
        s0 = peg$FAILED;
      }
    } else {
      peg$currPos = s0;
      s0 = peg$FAILED;
    }
    if (s0 === peg$FAILED) {
      s0 = peg$currPos;
      s1 = peg$parseNOT();
      if (s1 !== peg$FAILED) {
        s2 = peg$parse_();
        s3 = peg$parseprimaryExpression();
        if (s3 !== peg$FAILED) {
          peg$savedPos = s0;
          s0 = peg$f20(s3);
        } else {
          peg$currPos = s0;
          s0 = peg$FAILED;
        }
      } else {
        peg$currPos = s0;
        s0 = peg$FAILED;
      }
      if (s0 === peg$FAILED) {
        s0 = peg$parsecomparisonExpression();
      }
    }
    return s0;
  }
  function peg$parsecomparisonExpression() {
    let s0, s1, s2, s3, s4, s5;
    s0 = peg$currPos;
    s1 = peg$parsefilterarg();
    if (s1 !== peg$FAILED) {
      s2 = peg$parse_();
      s3 = peg$parsecomparisonOperator();
      if (s3 !== peg$FAILED) {
        s4 = peg$parse_();
        s5 = peg$parsefilterarg();
        if (s5 !== peg$FAILED) {
          peg$savedPos = s0;
          s0 = peg$f21(s1, s3, s5);
        } else {
          peg$currPos = s0;
          s0 = peg$FAILED;
        }
      } else {
        peg$currPos = s0;
        s0 = peg$FAILED;
      }
    } else {
      peg$currPos = s0;
      s0 = peg$FAILED;
    }
    if (s0 === peg$FAILED) {
      s0 = peg$currPos;
      s1 = peg$parsefilterarg();
      if (s1 !== peg$FAILED) {
        s2 = peg$parse_();
        s3 = peg$parseREG();
        if (s3 !== peg$FAILED) {
          s4 = peg$parse_();
          s5 = peg$parseregex();
          if (s5 !== peg$FAILED) {
            peg$savedPos = s0;
            s0 = peg$f22(s1, s5);
          } else {
            peg$currPos = s0;
            s0 = peg$FAILED;
          }
        } else {
          peg$currPos = s0;
          s0 = peg$FAILED;
        }
      } else {
        peg$currPos = s0;
        s0 = peg$FAILED;
      }
      if (s0 === peg$FAILED) {
        s0 = peg$currPos;
        s1 = peg$parsefilterarg();
        if (s1 !== peg$FAILED) {
          s2 = peg$parse_();
          s3 = peg$parseEMPT();
          if (s3 !== peg$FAILED) {
            peg$savedPos = s0;
            s0 = peg$f23(s1);
          } else {
            peg$currPos = s0;
            s0 = peg$FAILED;
          }
        } else {
          peg$currPos = s0;
          s0 = peg$FAILED;
        }
        if (s0 === peg$FAILED) {
          s0 = peg$parseprimaryExpression();
        }
      }
    }
    return s0;
  }
  function peg$parsecomparisonOperator() {
    let s0, s1;
    s0 = peg$currPos;
    s1 = peg$parseEQ();
    if (s1 !== peg$FAILED) {
      peg$savedPos = s0;
      s1 = peg$f24();
    }
    s0 = s1;
    if (s0 === peg$FAILED) {
      s0 = peg$currPos;
      s1 = peg$parseNE();
      if (s1 !== peg$FAILED) {
        peg$savedPos = s0;
        s1 = peg$f25();
      }
      s0 = s1;
      if (s0 === peg$FAILED) {
        s0 = peg$currPos;
        s1 = peg$parseLE();
        if (s1 !== peg$FAILED) {
          peg$savedPos = s0;
          s1 = peg$f26();
        }
        s0 = s1;
        if (s0 === peg$FAILED) {
          s0 = peg$currPos;
          s1 = peg$parseLT();
          if (s1 !== peg$FAILED) {
            peg$savedPos = s0;
            s1 = peg$f27();
          }
          s0 = s1;
          if (s0 === peg$FAILED) {
            s0 = peg$currPos;
            s1 = peg$parseGE();
            if (s1 !== peg$FAILED) {
              peg$savedPos = s0;
              s1 = peg$f28();
            }
            s0 = s1;
            if (s0 === peg$FAILED) {
              s0 = peg$currPos;
              s1 = peg$parseGT();
              if (s1 !== peg$FAILED) {
                peg$savedPos = s0;
                s1 = peg$f29();
              }
              s0 = s1;
              if (s0 === peg$FAILED) {
                s0 = peg$currPos;
                s1 = peg$parseIN();
                if (s1 !== peg$FAILED) {
                  peg$savedPos = s0;
                  s1 = peg$f30();
                }
                s0 = s1;
                if (s0 === peg$FAILED) {
                  s0 = peg$currPos;
                  s1 = peg$parseNIN();
                  if (s1 !== peg$FAILED) {
                    peg$savedPos = s0;
                    s1 = peg$f31();
                  }
                  s0 = s1;
                  if (s0 === peg$FAILED) {
                    s0 = peg$currPos;
                    s1 = peg$parseSUB();
                    if (s1 !== peg$FAILED) {
                      peg$savedPos = s0;
                      s1 = peg$f32();
                    }
                    s0 = s1;
                    if (s0 === peg$FAILED) {
                      s0 = peg$currPos;
                      s1 = peg$parseANY();
                      if (s1 !== peg$FAILED) {
                        peg$savedPos = s0;
                        s1 = peg$f33();
                      }
                      s0 = s1;
                      if (s0 === peg$FAILED) {
                        s0 = peg$currPos;
                        s1 = peg$parseNON();
                        if (s1 !== peg$FAILED) {
                          peg$savedPos = s0;
                          s1 = peg$f34();
                        }
                        s0 = s1;
                        if (s0 === peg$FAILED) {
                          s0 = peg$currPos;
                          s1 = peg$parseSIZO();
                          if (s1 !== peg$FAILED) {
                            peg$savedPos = s0;
                            s1 = peg$f35();
                          }
                          s0 = s1;
                          if (s0 === peg$FAILED) {
                            s0 = peg$currPos;
                            s1 = peg$parseSIZ();
                            if (s1 !== peg$FAILED) {
                              peg$savedPos = s0;
                              s1 = peg$f36();
                            }
                            s0 = s1;
                          }
                        }
                      }
                    }
                  }
                }
              }
            }
          }
        }
      }
    }
    return s0;
  }
  function peg$parseprimaryExpression() {
    let s0, s1, s2, s3, s4, s5;
    s0 = peg$currPos;
    s1 = peg$parsePAREN_LEFT();
    if (s1 !== peg$FAILED) {
      s2 = peg$parse_();
      s3 = peg$parseorExpression();
      if (s3 !== peg$FAILED) {
        s4 = peg$parse_();
        s5 = peg$parsePAREN_RIGHT();
        if (s5 !== peg$FAILED) {
          peg$savedPos = s0;
          s0 = peg$f37(s3);
        } else {
          peg$currPos = s0;
          s0 = peg$FAILED;
        }
      } else {
        peg$currPos = s0;
        s0 = peg$FAILED;
      }
    } else {
      peg$currPos = s0;
      s0 = peg$FAILED;
    }
    if (s0 === peg$FAILED) {
      s0 = peg$parsefunctionCall();
      if (s0 === peg$FAILED) {
        s0 = peg$parsefilterpath();
        if (s0 === peg$FAILED) {
          s0 = peg$currPos;
          s1 = peg$parseTRUE();
          if (s1 !== peg$FAILED) {
            peg$savedPos = s0;
            s1 = peg$f38();
          }
          s0 = s1;
          if (s0 === peg$FAILED) {
            s0 = peg$currPos;
            s1 = peg$parseFALSE();
            if (s1 !== peg$FAILED) {
              peg$savedPos = s0;
              s1 = peg$f39();
            }
            s0 = s1;
            if (s0 === peg$FAILED) {
              s0 = peg$currPos;
              s1 = peg$parseNULL();
              if (s1 !== peg$FAILED) {
                peg$savedPos = s0;
                s1 = peg$f40();
              }
              s0 = s1;
            }
          }
        }
      }
    }
    return s0;
  }
  function peg$parsefunctionCall() {
    let s0, s1, s2, s3, s4, s5, s6, s7;
    s0 = peg$currPos;
    s1 = peg$parsefunctionName();
    if (s1 !== peg$FAILED) {
      s2 = peg$parse_();
      s3 = peg$parsePAREN_LEFT();
      if (s3 !== peg$FAILED) {
        s4 = peg$parse_();
        s5 = peg$parsefunctionArgs();
        if (s5 === peg$FAILED) {
          s5 = null;
        }
        s6 = peg$parse_();
        s7 = peg$parsePAREN_RIGHT();
        if (s7 !== peg$FAILED) {
          peg$savedPos = s0;
          s0 = peg$f41(s1, s5);
        } else {
          peg$currPos = s0;
          s0 = peg$FAILED;
        }
      } else {
        peg$currPos = s0;
        s0 = peg$FAILED;
      }
    } else {
      peg$currPos = s0;
      s0 = peg$FAILED;
    }
    return s0;
  }
  function peg$parsefunctionName() {
    let s0, s1;
    s0 = peg$currPos;
    s1 = input.substr(peg$currPos, 6);
    if (s1.toLowerCase() === peg$c0) {
      peg$currPos += 6;
    } else {
      s1 = peg$FAILED;
      if (peg$silentFails === 0) {
        peg$fail(peg$e0);
      }
    }
    if (s1 !== peg$FAILED) {
      peg$savedPos = s0;
      s1 = peg$f42();
    }
    s0 = s1;
    if (s0 === peg$FAILED) {
      s0 = peg$currPos;
      s1 = input.substr(peg$currPos, 5);
      if (s1.toLowerCase() === peg$c1) {
        peg$currPos += 5;
      } else {
        s1 = peg$FAILED;
        if (peg$silentFails === 0) {
          peg$fail(peg$e1);
        }
      }
      if (s1 !== peg$FAILED) {
        peg$savedPos = s0;
        s1 = peg$f43();
      }
      s0 = s1;
      if (s0 === peg$FAILED) {
        s0 = peg$currPos;
        s1 = input.substr(peg$currPos, 5);
        if (s1.toLowerCase() === peg$c2) {
          peg$currPos += 5;
        } else {
          s1 = peg$FAILED;
          if (peg$silentFails === 0) {
            peg$fail(peg$e2);
          }
        }
        if (s1 !== peg$FAILED) {
          peg$savedPos = s0;
          s1 = peg$f44();
        }
        s0 = s1;
        if (s0 === peg$FAILED) {
          s0 = peg$currPos;
          s1 = input.substr(peg$currPos, 6);
          if (s1.toLowerCase() === peg$c3) {
            peg$currPos += 6;
          } else {
            s1 = peg$FAILED;
            if (peg$silentFails === 0) {
              peg$fail(peg$e3);
            }
          }
          if (s1 !== peg$FAILED) {
            peg$savedPos = s0;
            s1 = peg$f45();
          }
          s0 = s1;
          if (s0 === peg$FAILED) {
            s0 = peg$currPos;
            s1 = input.substr(peg$currPos, 5);
            if (s1.toLowerCase() === peg$c4) {
              peg$currPos += 5;
            } else {
              s1 = peg$FAILED;
              if (peg$silentFails === 0) {
                peg$fail(peg$e4);
              }
            }
            if (s1 !== peg$FAILED) {
              peg$savedPos = s0;
              s1 = peg$f46();
            }
            s0 = s1;
          }
        }
      }
    }
    return s0;
  }
  function peg$parsefunctionArgs() {
    let s0, s1, s2, s3, s4, s5, s6, s7;
    s0 = peg$currPos;
    s1 = peg$parsefunctionArg();
    if (s1 !== peg$FAILED) {
      s2 = [];
      s3 = peg$currPos;
      s4 = peg$parse_();
      s5 = peg$parseCOMMA();
      if (s5 !== peg$FAILED) {
        s6 = peg$parse_();
        s7 = peg$parsefunctionArg();
        if (s7 !== peg$FAILED) {
          s4 = [s4, s5, s6, s7];
          s3 = s4;
        } else {
          peg$currPos = s3;
          s3 = peg$FAILED;
        }
      } else {
        peg$currPos = s3;
        s3 = peg$FAILED;
      }
      while (s3 !== peg$FAILED) {
        s2.push(s3);
        s3 = peg$currPos;
        s4 = peg$parse_();
        s5 = peg$parseCOMMA();
        if (s5 !== peg$FAILED) {
          s6 = peg$parse_();
          s7 = peg$parsefunctionArg();
          if (s7 !== peg$FAILED) {
            s4 = [s4, s5, s6, s7];
            s3 = s4;
          } else {
            peg$currPos = s3;
            s3 = peg$FAILED;
          }
        } else {
          peg$currPos = s3;
          s3 = peg$FAILED;
        }
      }
      peg$savedPos = s0;
      s0 = peg$f47(s1, s2);
    } else {
      peg$currPos = s0;
      s0 = peg$FAILED;
    }
    return s0;
  }
  function peg$parsefunctionArg() {
    let s0;
    s0 = peg$parsefilterarg();
    if (s0 === peg$FAILED) {
      s0 = peg$parsestringLiteral();
    }
    return s0;
  }
  function peg$parsefilterarg() {
    let s0, s1, s2, s3, s4, s5, s6, s7;
    s0 = peg$currPos;
    s1 = peg$parsefilterargMulti();
    if (s1 !== peg$FAILED) {
      s2 = [];
      s3 = peg$currPos;
      s4 = peg$parse_();
      s5 = input.charAt(peg$currPos);
      if (peg$r0.test(s5)) {
        peg$currPos++;
      } else {
        s5 = peg$FAILED;
        if (peg$silentFails === 0) {
          peg$fail(peg$e5);
        }
      }
      if (s5 !== peg$FAILED) {
        s6 = peg$parse_();
        s7 = peg$parsefilterargMulti();
        if (s7 !== peg$FAILED) {
          s4 = [s4, s5, s6, s7];
          s3 = s4;
        } else {
          peg$currPos = s3;
          s3 = peg$FAILED;
        }
      } else {
        peg$currPos = s3;
        s3 = peg$FAILED;
      }
      while (s3 !== peg$FAILED) {
        s2.push(s3);
        s3 = peg$currPos;
        s4 = peg$parse_();
        s5 = input.charAt(peg$currPos);
        if (peg$r0.test(s5)) {
          peg$currPos++;
        } else {
          s5 = peg$FAILED;
          if (peg$silentFails === 0) {
            peg$fail(peg$e5);
          }
        }
        if (s5 !== peg$FAILED) {
          s6 = peg$parse_();
          s7 = peg$parsefilterargMulti();
          if (s7 !== peg$FAILED) {
            s4 = [s4, s5, s6, s7];
            s3 = s4;
          } else {
            peg$currPos = s3;
            s3 = peg$FAILED;
          }
        } else {
          peg$currPos = s3;
          s3 = peg$FAILED;
        }
      }
      peg$savedPos = s0;
      s0 = peg$f48(s1, s2);
    } else {
      peg$currPos = s0;
      s0 = peg$FAILED;
    }
    return s0;
  }
  function peg$parsefilterargMulti() {
    let s0, s1, s2, s3, s4, s5, s6, s7;
    s0 = peg$currPos;
    s1 = peg$parsefilterargUnary();
    if (s1 !== peg$FAILED) {
      s2 = [];
      s3 = peg$currPos;
      s4 = peg$parse_();
      s5 = input.charAt(peg$currPos);
      if (peg$r1.test(s5)) {
        peg$currPos++;
      } else {
        s5 = peg$FAILED;
        if (peg$silentFails === 0) {
          peg$fail(peg$e6);
        }
      }
      if (s5 !== peg$FAILED) {
        s6 = peg$parse_();
        s7 = peg$parsefilterargUnary();
        if (s7 !== peg$FAILED) {
          s4 = [s4, s5, s6, s7];
          s3 = s4;
        } else {
          peg$currPos = s3;
          s3 = peg$FAILED;
        }
      } else {
        peg$currPos = s3;
        s3 = peg$FAILED;
      }
      while (s3 !== peg$FAILED) {
        s2.push(s3);
        s3 = peg$currPos;
        s4 = peg$parse_();
        s5 = input.charAt(peg$currPos);
        if (peg$r1.test(s5)) {
          peg$currPos++;
        } else {
          s5 = peg$FAILED;
          if (peg$silentFails === 0) {
            peg$fail(peg$e6);
          }
        }
        if (s5 !== peg$FAILED) {
          s6 = peg$parse_();
          s7 = peg$parsefilterargUnary();
          if (s7 !== peg$FAILED) {
            s4 = [s4, s5, s6, s7];
            s3 = s4;
          } else {
            peg$currPos = s3;
            s3 = peg$FAILED;
          }
        } else {
          peg$currPos = s3;
          s3 = peg$FAILED;
        }
      }
      peg$savedPos = s0;
      s0 = peg$f49(s1, s2);
    } else {
      peg$currPos = s0;
      s0 = peg$FAILED;
    }
    return s0;
  }
  function peg$parsefilterargUnary() {
    let s0, s1, s2, s3, s4, s5;
    s0 = peg$currPos;
    s1 = peg$parsePAREN_LEFT();
    if (s1 !== peg$FAILED) {
      s2 = peg$parse_();
      s3 = peg$parsefilterarg();
      if (s3 !== peg$FAILED) {
        s4 = peg$parse_();
        s5 = peg$parsePAREN_RIGHT();
        if (s5 !== peg$FAILED) {
          peg$savedPos = s0;
          s0 = peg$f50(s3);
        } else {
          peg$currPos = s0;
          s0 = peg$FAILED;
        }
      } else {
        peg$currPos = s0;
        s0 = peg$FAILED;
      }
    } else {
      peg$currPos = s0;
      s0 = peg$FAILED;
    }
    if (s0 === peg$FAILED) {
      s0 = peg$parsefunctionCall();
      if (s0 === peg$FAILED) {
        s0 = peg$parsevalue();
        if (s0 === peg$FAILED) {
          s0 = peg$parsefilterpath();
        }
      }
    }
    return s0;
  }
  function peg$parsefilterpath() {
    let s0, s1, s2, s3;
    s0 = peg$currPos;
    s1 = peg$parseROOT_VALUE();
    if (s1 !== peg$FAILED) {
      s2 = [];
      s3 = peg$parsesegment();
      while (s3 !== peg$FAILED) {
        s2.push(s3);
        s3 = peg$parsesegment();
      }
      peg$savedPos = s0;
      s0 = peg$f51(s2);
    } else {
      peg$currPos = s0;
      s0 = peg$FAILED;
    }
    if (s0 === peg$FAILED) {
      s0 = peg$currPos;
      s1 = peg$parseCURRENT_VALUE();
      if (s1 !== peg$FAILED) {
        s2 = [];
        s3 = peg$parsesegment();
        while (s3 !== peg$FAILED) {
          s2.push(s3);
          s3 = peg$parsesegment();
        }
        peg$savedPos = s0;
        s0 = peg$f52(s2);
      } else {
        peg$currPos = s0;
        s0 = peg$FAILED;
      }
    }
    return s0;
  }
  function peg$parseslices() {
    let s0, s1, s2, s3, s4, s5, s6;
    s0 = peg$currPos;
    s1 = peg$parseNUMBER();
    if (s1 === peg$FAILED) {
      s1 = null;
    }
    s2 = peg$parseCOLON();
    if (s2 !== peg$FAILED) {
      s3 = peg$parseNUMBER();
      if (s3 === peg$FAILED) {
        s3 = null;
      }
      s4 = peg$currPos;
      s5 = peg$parseCOLON();
      if (s5 !== peg$FAILED) {
        s6 = peg$parseNUMBER();
        if (s6 === peg$FAILED) {
          s6 = null;
        }
        s5 = [s5, s6];
        s4 = s5;
      } else {
        peg$currPos = s4;
        s4 = peg$FAILED;
      }
      if (s4 === peg$FAILED) {
        s4 = null;
      }
      peg$savedPos = s0;
      s0 = peg$f53(s1, s3, s4);
    } else {
      peg$currPos = s0;
      s0 = peg$FAILED;
    }
    return s0;
  }
  function peg$parseselectorList() {
    let s0, s1, s2, s3, s4, s5, s6, s7;
    s0 = peg$currPos;
    s1 = peg$parseselector();
    if (s1 !== peg$FAILED) {
      s2 = [];
      s3 = peg$currPos;
      s4 = peg$parse_();
      s5 = peg$parseCOMMA();
      if (s5 !== peg$FAILED) {
        s6 = peg$parse_();
        s7 = peg$parseselector();
        if (s7 !== peg$FAILED) {
          s4 = [s4, s5, s6, s7];
          s3 = s4;
        } else {
          peg$currPos = s3;
          s3 = peg$FAILED;
        }
      } else {
        peg$currPos = s3;
        s3 = peg$FAILED;
      }
      if (s3 !== peg$FAILED) {
        while (s3 !== peg$FAILED) {
          s2.push(s3);
          s3 = peg$currPos;
          s4 = peg$parse_();
          s5 = peg$parseCOMMA();
          if (s5 !== peg$FAILED) {
            s6 = peg$parse_();
            s7 = peg$parseselector();
            if (s7 !== peg$FAILED) {
              s4 = [s4, s5, s6, s7];
              s3 = s4;
            } else {
              peg$currPos = s3;
              s3 = peg$FAILED;
            }
          } else {
            peg$currPos = s3;
            s3 = peg$FAILED;
          }
        }
      } else {
        s2 = peg$FAILED;
      }
      if (s2 !== peg$FAILED) {
        peg$savedPos = s0;
        s0 = peg$f54(s1, s2);
      } else {
        peg$currPos = s0;
        s0 = peg$FAILED;
      }
    } else {
      peg$currPos = s0;
      s0 = peg$FAILED;
    }
    return s0;
  }
  function peg$parseselector() {
    let s0, s1;
    s0 = peg$currPos;
    s1 = peg$parseSTAR();
    if (s1 !== peg$FAILED) {
      peg$savedPos = s0;
      s1 = peg$f55();
    }
    s0 = s1;
    if (s0 === peg$FAILED) {
      s0 = peg$currPos;
      s1 = peg$parseNUMBER();
      if (s1 !== peg$FAILED) {
        peg$savedPos = s0;
        s1 = peg$f56(s1);
      }
      s0 = s1;
      if (s0 === peg$FAILED) {
        s0 = peg$currPos;
        s1 = peg$parseSTRING();
        if (s1 !== peg$FAILED) {
          peg$savedPos = s0;
          s1 = peg$f57(s1);
        }
        s0 = s1;
        if (s0 === peg$FAILED) {
          s0 = peg$currPos;
          s1 = peg$parseidentifier();
          if (s1 !== peg$FAILED) {
            peg$savedPos = s0;
            s1 = peg$f58(s1);
          }
          s0 = s1;
        }
      }
    }
    return s0;
  }
  function peg$parseregex() {
    let s0, s1, s2;
    s0 = peg$currPos;
    s1 = peg$parseREGEX_EXPR();
    if (s1 !== peg$FAILED) {
      s2 = peg$parseREGEX_OPT();
      peg$savedPos = s0;
      s0 = peg$f59(s1, s2);
    } else {
      peg$currPos = s0;
      s0 = peg$FAILED;
    }
    return s0;
  }
  function peg$parsevalue() {
    let s0, s1;
    s0 = peg$currPos;
    s1 = peg$parsestringLiteral();
    if (s1 !== peg$FAILED) {
      peg$savedPos = s0;
      s1 = peg$f60(s1);
    }
    s0 = s1;
    if (s0 === peg$FAILED) {
      s0 = peg$currPos;
      s1 = peg$parseNUMBER();
      if (s1 !== peg$FAILED) {
        peg$savedPos = s0;
        s1 = peg$f61(s1);
      }
      s0 = s1;
      if (s0 === peg$FAILED) {
        s0 = peg$parseobj();
        if (s0 === peg$FAILED) {
          s0 = peg$parsearray();
          if (s0 === peg$FAILED) {
            s0 = peg$currPos;
            s1 = peg$parseTRUE();
            if (s1 !== peg$FAILED) {
              peg$savedPos = s0;
              s1 = peg$f62();
            }
            s0 = s1;
            if (s0 === peg$FAILED) {
              s0 = peg$currPos;
              s1 = peg$parseFALSE();
              if (s1 !== peg$FAILED) {
                peg$savedPos = s0;
                s1 = peg$f63();
              }
              s0 = s1;
              if (s0 === peg$FAILED) {
                s0 = peg$currPos;
                s1 = peg$parseNULL();
                if (s1 !== peg$FAILED) {
                  peg$savedPos = s0;
                  s1 = peg$f64();
                }
                s0 = s1;
              }
            }
          }
        }
      }
    }
    return s0;
  }
  function peg$parsestringLiteral() {
    let s0, s1;
    peg$silentFails++;
    s0 = peg$currPos;
    s1 = peg$parseSTRING();
    if (s1 !== peg$FAILED) {
      peg$savedPos = s0;
      s1 = peg$f65(s1);
    }
    s0 = s1;
    peg$silentFails--;
    if (s0 === peg$FAILED) {
      s1 = peg$FAILED;
      if (peg$silentFails === 0) {
        peg$fail(peg$e7);
      }
    }
    return s0;
  }
  function peg$parseobj() {
    let s0, s1, s2, s3, s4, s5;
    s0 = peg$currPos;
    s1 = peg$parseBRACE_LEFT();
    if (s1 !== peg$FAILED) {
      s2 = peg$parse_();
      s3 = peg$parsepairList();
      if (s3 === peg$FAILED) {
        s3 = null;
      }
      s4 = peg$parse_();
      s5 = peg$parseBRACE_RIGHT();
      if (s5 !== peg$FAILED) {
        peg$savedPos = s0;
        s0 = peg$f66(s3);
      } else {
        peg$currPos = s0;
        s0 = peg$FAILED;
      }
    } else {
      peg$currPos = s0;
      s0 = peg$FAILED;
    }
    return s0;
  }
  function peg$parsepairList() {
    let s0, s1, s2, s3, s4, s5, s6, s7;
    s0 = peg$currPos;
    s1 = peg$parsepair();
    if (s1 !== peg$FAILED) {
      s2 = [];
      s3 = peg$currPos;
      s4 = peg$parse_();
      s5 = peg$parseCOMMA();
      if (s5 !== peg$FAILED) {
        s6 = peg$parse_();
        s7 = peg$parsepair();
        if (s7 !== peg$FAILED) {
          s4 = [s4, s5, s6, s7];
          s3 = s4;
        } else {
          peg$currPos = s3;
          s3 = peg$FAILED;
        }
      } else {
        peg$currPos = s3;
        s3 = peg$FAILED;
      }
      while (s3 !== peg$FAILED) {
        s2.push(s3);
        s3 = peg$currPos;
        s4 = peg$parse_();
        s5 = peg$parseCOMMA();
        if (s5 !== peg$FAILED) {
          s6 = peg$parse_();
          s7 = peg$parsepair();
          if (s7 !== peg$FAILED) {
            s4 = [s4, s5, s6, s7];
            s3 = s4;
          } else {
            peg$currPos = s3;
            s3 = peg$FAILED;
          }
        } else {
          peg$currPos = s3;
          s3 = peg$FAILED;
        }
      }
      peg$savedPos = s0;
      s0 = peg$f67(s1, s2);
    } else {
      peg$currPos = s0;
      s0 = peg$FAILED;
    }
    return s0;
  }
  function peg$parsepair() {
    let s0, s1, s2, s3, s4, s5;
    s0 = peg$currPos;
    s1 = peg$parseSTRING();
    if (s1 !== peg$FAILED) {
      s2 = peg$parse_();
      s3 = peg$parseCOLON();
      if (s3 !== peg$FAILED) {
        s4 = peg$parse_();
        s5 = peg$parsevalue();
        if (s5 !== peg$FAILED) {
          peg$savedPos = s0;
          s0 = peg$f68(s1, s5);
        } else {
          peg$currPos = s0;
          s0 = peg$FAILED;
        }
      } else {
        peg$currPos = s0;
        s0 = peg$FAILED;
      }
    } else {
      peg$currPos = s0;
      s0 = peg$FAILED;
    }
    return s0;
  }
  function peg$parsearray() {
    let s0, s1, s2, s3, s4, s5;
    s0 = peg$currPos;
    s1 = peg$parseBRACKET_LEFT();
    if (s1 !== peg$FAILED) {
      s2 = peg$parse_();
      s3 = peg$parsevalueList();
      if (s3 === peg$FAILED) {
        s3 = null;
      }
      s4 = peg$parse_();
      s5 = peg$parseBRACKET_RIGHT();
      if (s5 !== peg$FAILED) {
        peg$savedPos = s0;
        s0 = peg$f69(s3);
      } else {
        peg$currPos = s0;
        s0 = peg$FAILED;
      }
    } else {
      peg$currPos = s0;
      s0 = peg$FAILED;
    }
    return s0;
  }
  function peg$parsevalueList() {
    let s0, s1, s2, s3, s4, s5, s6, s7;
    s0 = peg$currPos;
    s1 = peg$parsevalue();
    if (s1 !== peg$FAILED) {
      s2 = [];
      s3 = peg$currPos;
      s4 = peg$parse_();
      s5 = peg$parseCOMMA();
      if (s5 !== peg$FAILED) {
        s6 = peg$parse_();
        s7 = peg$parsevalue();
        if (s7 !== peg$FAILED) {
          s4 = [s4, s5, s6, s7];
          s3 = s4;
        } else {
          peg$currPos = s3;
          s3 = peg$FAILED;
        }
      } else {
        peg$currPos = s3;
        s3 = peg$FAILED;
      }
      while (s3 !== peg$FAILED) {
        s2.push(s3);
        s3 = peg$currPos;
        s4 = peg$parse_();
        s5 = peg$parseCOMMA();
        if (s5 !== peg$FAILED) {
          s6 = peg$parse_();
          s7 = peg$parsevalue();
          if (s7 !== peg$FAILED) {
            s4 = [s4, s5, s6, s7];
            s3 = s4;
          } else {
            peg$currPos = s3;
            s3 = peg$FAILED;
          }
        } else {
          peg$currPos = s3;
          s3 = peg$FAILED;
        }
      }
      peg$savedPos = s0;
      s0 = peg$f70(s1, s2);
    } else {
      peg$currPos = s0;
      s0 = peg$FAILED;
    }
    return s0;
  }
  function peg$parseidentifier() {
    let s0, s1;
    s0 = peg$currPos;
    s1 = peg$parseKEY();
    if (s1 !== peg$FAILED) {
      peg$savedPos = s0;
      s1 = peg$f71(s1);
    }
    s0 = s1;
    if (s0 === peg$FAILED) {
      s0 = peg$currPos;
      s1 = peg$parseSPECIAL_KEY();
      if (s1 !== peg$FAILED) {
        peg$savedPos = s0;
        s1 = peg$f72(s1);
      }
      s0 = s1;
      if (s0 === peg$FAILED) {
        s0 = peg$currPos;
        s1 = peg$parseTRUE();
        if (s1 !== peg$FAILED) {
          peg$savedPos = s0;
          s1 = peg$f73();
        }
        s0 = s1;
        if (s0 === peg$FAILED) {
          s0 = peg$currPos;
          s1 = peg$parseFALSE();
          if (s1 !== peg$FAILED) {
            peg$savedPos = s0;
            s1 = peg$f74();
          }
          s0 = s1;
          if (s0 === peg$FAILED) {
            s0 = peg$currPos;
            s1 = peg$parseNULL();
            if (s1 !== peg$FAILED) {
              peg$savedPos = s0;
              s1 = peg$f75();
            }
            s0 = s1;
          }
        }
      }
    }
    return s0;
  }
  function peg$parseROOT_VALUE() {
    let s0;
    if (input.charCodeAt(peg$currPos) === 36) {
      s0 = peg$c5;
      peg$currPos++;
    } else {
      s0 = peg$FAILED;
      if (peg$silentFails === 0) {
        peg$fail(peg$e8);
      }
    }
    return s0;
  }
  function peg$parseCURRENT_VALUE() {
    let s0;
    if (input.charCodeAt(peg$currPos) === 64) {
      s0 = peg$c6;
      peg$currPos++;
    } else {
      s0 = peg$FAILED;
      if (peg$silentFails === 0) {
        peg$fail(peg$e9);
      }
    }
    return s0;
  }
  function peg$parseDOTDOT() {
    let s0;
    if (input.substr(peg$currPos, 2) === peg$c7) {
      s0 = peg$c7;
      peg$currPos += 2;
    } else {
      s0 = peg$FAILED;
      if (peg$silentFails === 0) {
        peg$fail(peg$e10);
      }
    }
    return s0;
  }
  function peg$parseDOT() {
    let s0;
    if (input.charCodeAt(peg$currPos) === 46) {
      s0 = peg$c8;
      peg$currPos++;
    } else {
      s0 = peg$FAILED;
      if (peg$silentFails === 0) {
        peg$fail(peg$e11);
      }
    }
    return s0;
  }
  function peg$parseSTAR() {
    let s0;
    if (input.charCodeAt(peg$currPos) === 42) {
      s0 = peg$c9;
      peg$currPos++;
    } else {
      s0 = peg$FAILED;
      if (peg$silentFails === 0) {
        peg$fail(peg$e12);
      }
    }
    return s0;
  }
  function peg$parseAND() {
    let s0;
    if (input.substr(peg$currPos, 2) === peg$c10) {
      s0 = peg$c10;
      peg$currPos += 2;
    } else {
      s0 = peg$FAILED;
      if (peg$silentFails === 0) {
        peg$fail(peg$e13);
      }
    }
    return s0;
  }
  function peg$parseOR() {
    let s0;
    if (input.substr(peg$currPos, 2) === peg$c11) {
      s0 = peg$c11;
      peg$currPos += 2;
    } else {
      s0 = peg$FAILED;
      if (peg$silentFails === 0) {
        peg$fail(peg$e14);
      }
    }
    return s0;
  }
  function peg$parseNOT() {
    let s0;
    if (input.charCodeAt(peg$currPos) === 33) {
      s0 = peg$c12;
      peg$currPos++;
    } else {
      s0 = peg$FAILED;
      if (peg$silentFails === 0) {
        peg$fail(peg$e15);
      }
    }
    return s0;
  }
  function peg$parseEQ() {
    let s0;
    if (input.substr(peg$currPos, 2) === peg$c13) {
      s0 = peg$c13;
      peg$currPos += 2;
    } else {
      s0 = peg$FAILED;
      if (peg$silentFails === 0) {
        peg$fail(peg$e16);
      }
    }
    return s0;
  }
  function peg$parseNE() {
    let s0;
    if (input.substr(peg$currPos, 2) === peg$c14) {
      s0 = peg$c14;
      peg$currPos += 2;
    } else {
      s0 = peg$FAILED;
      if (peg$silentFails === 0) {
        peg$fail(peg$e17);
      }
    }
    return s0;
  }
  function peg$parseGE() {
    let s0;
    if (input.substr(peg$currPos, 2) === peg$c15) {
      s0 = peg$c15;
      peg$currPos += 2;
    } else {
      s0 = peg$FAILED;
      if (peg$silentFails === 0) {
        peg$fail(peg$e18);
      }
    }
    return s0;
  }
  function peg$parseGT() {
    let s0;
    if (input.charCodeAt(peg$currPos) === 62) {
      s0 = peg$c16;
      peg$currPos++;
    } else {
      s0 = peg$FAILED;
      if (peg$silentFails === 0) {
        peg$fail(peg$e19);
      }
    }
    return s0;
  }
  function peg$parseLE() {
    let s0;
    if (input.substr(peg$currPos, 2) === peg$c17) {
      s0 = peg$c17;
      peg$currPos += 2;
    } else {
      s0 = peg$FAILED;
      if (peg$silentFails === 0) {
        peg$fail(peg$e20);
      }
    }
    return s0;
  }
  function peg$parseLT() {
    let s0;
    if (input.charCodeAt(peg$currPos) === 60) {
      s0 = peg$c18;
      peg$currPos++;
    } else {
      s0 = peg$FAILED;
      if (peg$silentFails === 0) {
        peg$fail(peg$e21);
      }
    }
    return s0;
  }
  function peg$parseREG() {
    let s0;
    if (input.substr(peg$currPos, 2) === peg$c19) {
      s0 = peg$c19;
      peg$currPos += 2;
    } else {
      s0 = peg$FAILED;
      if (peg$silentFails === 0) {
        peg$fail(peg$e22);
      }
    }
    return s0;
  }
  function peg$parseIN() {
    let s0, s1, s2, s3;
    s0 = peg$currPos;
    if (input.substr(peg$currPos, 2) === peg$c20) {
      s1 = peg$c20;
      peg$currPos += 2;
    } else {
      s1 = peg$FAILED;
      if (peg$silentFails === 0) {
        peg$fail(peg$e23);
      }
    }
    if (s1 !== peg$FAILED) {
      s2 = peg$currPos;
      peg$silentFails++;
      s3 = peg$parseidentifierChar();
      peg$silentFails--;
      if (s3 === peg$FAILED) {
        s2 = void 0;
      } else {
        peg$currPos = s2;
        s2 = peg$FAILED;
      }
      if (s2 !== peg$FAILED) {
        s1 = [s1, s2];
        s0 = s1;
      } else {
        peg$currPos = s0;
        s0 = peg$FAILED;
      }
    } else {
      peg$currPos = s0;
      s0 = peg$FAILED;
    }
    return s0;
  }
  function peg$parseNIN() {
    let s0, s1, s2, s3;
    s0 = peg$currPos;
    if (input.substr(peg$currPos, 3) === peg$c21) {
      s1 = peg$c21;
      peg$currPos += 3;
    } else {
      s1 = peg$FAILED;
      if (peg$silentFails === 0) {
        peg$fail(peg$e24);
      }
    }
    if (s1 !== peg$FAILED) {
      s2 = peg$currPos;
      peg$silentFails++;
      s3 = peg$parseidentifierChar();
      peg$silentFails--;
      if (s3 === peg$FAILED) {
        s2 = void 0;
      } else {
        peg$currPos = s2;
        s2 = peg$FAILED;
      }
      if (s2 !== peg$FAILED) {
        s1 = [s1, s2];
        s0 = s1;
      } else {
        peg$currPos = s0;
        s0 = peg$FAILED;
      }
    } else {
      peg$currPos = s0;
      s0 = peg$FAILED;
    }
    return s0;
  }
  function peg$parseSUB() {
    let s0, s1, s2, s3;
    s0 = peg$currPos;
    if (input.substr(peg$currPos, 8) === peg$c22) {
      s1 = peg$c22;
      peg$currPos += 8;
    } else {
      s1 = peg$FAILED;
      if (peg$silentFails === 0) {
        peg$fail(peg$e25);
      }
    }
    if (s1 !== peg$FAILED) {
      s2 = peg$currPos;
      peg$silentFails++;
      s3 = peg$parseidentifierChar();
      peg$silentFails--;
      if (s3 === peg$FAILED) {
        s2 = void 0;
      } else {
        peg$currPos = s2;
        s2 = peg$FAILED;
      }
      if (s2 !== peg$FAILED) {
        s1 = [s1, s2];
        s0 = s1;
      } else {
        peg$currPos = s0;
        s0 = peg$FAILED;
      }
    } else {
      peg$currPos = s0;
      s0 = peg$FAILED;
    }
    return s0;
  }
  function peg$parseANY() {
    let s0, s1, s2, s3;
    s0 = peg$currPos;
    if (input.substr(peg$currPos, 5) === peg$c23) {
      s1 = peg$c23;
      peg$currPos += 5;
    } else {
      s1 = peg$FAILED;
      if (peg$silentFails === 0) {
        peg$fail(peg$e26);
      }
    }
    if (s1 !== peg$FAILED) {
      s2 = peg$currPos;
      peg$silentFails++;
      s3 = peg$parseidentifierChar();
      peg$silentFails--;
      if (s3 === peg$FAILED) {
        s2 = void 0;
      } else {
        peg$currPos = s2;
        s2 = peg$FAILED;
      }
      if (s2 !== peg$FAILED) {
        s1 = [s1, s2];
        s0 = s1;
      } else {
        peg$currPos = s0;
        s0 = peg$FAILED;
      }
    } else {
      peg$currPos = s0;
      s0 = peg$FAILED;
    }
    return s0;
  }
  function peg$parseNON() {
    let s0, s1, s2, s3;
    s0 = peg$currPos;
    if (input.substr(peg$currPos, 6) === peg$c24) {
      s1 = peg$c24;
      peg$currPos += 6;
    } else {
      s1 = peg$FAILED;
      if (peg$silentFails === 0) {
        peg$fail(peg$e27);
      }
    }
    if (s1 !== peg$FAILED) {
      s2 = peg$currPos;
      peg$silentFails++;
      s3 = peg$parseidentifierChar();
      peg$silentFails--;
      if (s3 === peg$FAILED) {
        s2 = void 0;
      } else {
        peg$currPos = s2;
        s2 = peg$FAILED;
      }
      if (s2 !== peg$FAILED) {
        s1 = [s1, s2];
        s0 = s1;
      } else {
        peg$currPos = s0;
        s0 = peg$FAILED;
      }
    } else {
      peg$currPos = s0;
      s0 = peg$FAILED;
    }
    return s0;
  }
  function peg$parseSIZO() {
    let s0, s1, s2, s3;
    s0 = peg$currPos;
    if (input.substr(peg$currPos, 6) === peg$c25) {
      s1 = peg$c25;
      peg$currPos += 6;
    } else {
      s1 = peg$FAILED;
      if (peg$silentFails === 0) {
        peg$fail(peg$e28);
      }
    }
    if (s1 !== peg$FAILED) {
      s2 = peg$currPos;
      peg$silentFails++;
      s3 = peg$parseidentifierChar();
      peg$silentFails--;
      if (s3 === peg$FAILED) {
        s2 = void 0;
      } else {
        peg$currPos = s2;
        s2 = peg$FAILED;
      }
      if (s2 !== peg$FAILED) {
        s1 = [s1, s2];
        s0 = s1;
      } else {
        peg$currPos = s0;
        s0 = peg$FAILED;
      }
    } else {
      peg$currPos = s0;
      s0 = peg$FAILED;
    }
    return s0;
  }
  function peg$parseSIZ() {
    let s0, s1, s2, s3;
    s0 = peg$currPos;
    if (input.substr(peg$currPos, 4) === peg$c26) {
      s1 = peg$c26;
      peg$currPos += 4;
    } else {
      s1 = peg$FAILED;
      if (peg$silentFails === 0) {
        peg$fail(peg$e29);
      }
    }
    if (s1 !== peg$FAILED) {
      s2 = peg$currPos;
      peg$silentFails++;
      s3 = peg$parseidentifierChar();
      peg$silentFails--;
      if (s3 === peg$FAILED) {
        s2 = void 0;
      } else {
        peg$currPos = s2;
        s2 = peg$FAILED;
      }
      if (s2 !== peg$FAILED) {
        s1 = [s1, s2];
        s0 = s1;
      } else {
        peg$currPos = s0;
        s0 = peg$FAILED;
      }
    } else {
      peg$currPos = s0;
      s0 = peg$FAILED;
    }
    return s0;
  }
  function peg$parseEMPT() {
    let s0, s1, s2, s3;
    s0 = peg$currPos;
    if (input.substr(peg$currPos, 5) === peg$c27) {
      s1 = peg$c27;
      peg$currPos += 5;
    } else {
      s1 = peg$FAILED;
      if (peg$silentFails === 0) {
        peg$fail(peg$e30);
      }
    }
    if (s1 !== peg$FAILED) {
      s2 = peg$currPos;
      peg$silentFails++;
      s3 = peg$parseidentifierChar();
      peg$silentFails--;
      if (s3 === peg$FAILED) {
        s2 = void 0;
      } else {
        peg$currPos = s2;
        s2 = peg$FAILED;
      }
      if (s2 !== peg$FAILED) {
        s1 = [s1, s2];
        s0 = s1;
      } else {
        peg$currPos = s0;
        s0 = peg$FAILED;
      }
    } else {
      peg$currPos = s0;
      s0 = peg$FAILED;
    }
    return s0;
  }
  function peg$parseTRUE() {
    let s0, s1, s2, s3;
    s0 = peg$currPos;
    if (input.substr(peg$currPos, 4) === peg$c28) {
      s1 = peg$c28;
      peg$currPos += 4;
    } else {
      s1 = peg$FAILED;
      if (peg$silentFails === 0) {
        peg$fail(peg$e31);
      }
    }
    if (s1 !== peg$FAILED) {
      s2 = peg$currPos;
      peg$silentFails++;
      s3 = peg$parseidentifierChar();
      peg$silentFails--;
      if (s3 === peg$FAILED) {
        s2 = void 0;
      } else {
        peg$currPos = s2;
        s2 = peg$FAILED;
      }
      if (s2 !== peg$FAILED) {
        s1 = [s1, s2];
        s0 = s1;
      } else {
        peg$currPos = s0;
        s0 = peg$FAILED;
      }
    } else {
      peg$currPos = s0;
      s0 = peg$FAILED;
    }
    return s0;
  }
  function peg$parseFALSE() {
    let s0, s1, s2, s3;
    s0 = peg$currPos;
    if (input.substr(peg$currPos, 5) === peg$c29) {
      s1 = peg$c29;
      peg$currPos += 5;
    } else {
      s1 = peg$FAILED;
      if (peg$silentFails === 0) {
        peg$fail(peg$e32);
      }
    }
    if (s1 !== peg$FAILED) {
      s2 = peg$currPos;
      peg$silentFails++;
      s3 = peg$parseidentifierChar();
      peg$silentFails--;
      if (s3 === peg$FAILED) {
        s2 = void 0;
      } else {
        peg$currPos = s2;
        s2 = peg$FAILED;
      }
      if (s2 !== peg$FAILED) {
        s1 = [s1, s2];
        s0 = s1;
      } else {
        peg$currPos = s0;
        s0 = peg$FAILED;
      }
    } else {
      peg$currPos = s0;
      s0 = peg$FAILED;
    }
    return s0;
  }
  function peg$parseNULL() {
    let s0, s1, s2, s3;
    s0 = peg$currPos;
    if (input.substr(peg$currPos, 4) === peg$c30) {
      s1 = peg$c30;
      peg$currPos += 4;
    } else {
      s1 = peg$FAILED;
      if (peg$silentFails === 0) {
        peg$fail(peg$e33);
      }
    }
    if (s1 !== peg$FAILED) {
      s2 = peg$currPos;
      peg$silentFails++;
      s3 = peg$parseidentifierChar();
      peg$silentFails--;
      if (s3 === peg$FAILED) {
        s2 = void 0;
      } else {
        peg$currPos = s2;
        s2 = peg$FAILED;
      }
      if (s2 !== peg$FAILED) {
        s1 = [s1, s2];
        s0 = s1;
      } else {
        peg$currPos = s0;
        s0 = peg$FAILED;
      }
    } else {
      peg$currPos = s0;
      s0 = peg$FAILED;
    }
    return s0;
  }
  function peg$parseBRACE_LEFT() {
    let s0;
    if (input.charCodeAt(peg$currPos) === 123) {
      s0 = peg$c31;
      peg$currPos++;
    } else {
      s0 = peg$FAILED;
      if (peg$silentFails === 0) {
        peg$fail(peg$e34);
      }
    }
    return s0;
  }
  function peg$parseBRACE_RIGHT() {
    let s0;
    if (input.charCodeAt(peg$currPos) === 125) {
      s0 = peg$c32;
      peg$currPos++;
    } else {
      s0 = peg$FAILED;
      if (peg$silentFails === 0) {
        peg$fail(peg$e35);
      }
    }
    return s0;
  }
  function peg$parseBRACKET_LEFT() {
    let s0;
    if (input.charCodeAt(peg$currPos) === 91) {
      s0 = peg$c33;
      peg$currPos++;
    } else {
      s0 = peg$FAILED;
      if (peg$silentFails === 0) {
        peg$fail(peg$e36);
      }
    }
    return s0;
  }
  function peg$parseBRACKET_RIGHT() {
    let s0;
    if (input.charCodeAt(peg$currPos) === 93) {
      s0 = peg$c34;
      peg$currPos++;
    } else {
      s0 = peg$FAILED;
      if (peg$silentFails === 0) {
        peg$fail(peg$e37);
      }
    }
    return s0;
  }
  function peg$parseCOLON() {
    let s0;
    if (input.charCodeAt(peg$currPos) === 58) {
      s0 = peg$c35;
      peg$currPos++;
    } else {
      s0 = peg$FAILED;
      if (peg$silentFails === 0) {
        peg$fail(peg$e38);
      }
    }
    return s0;
  }
  function peg$parseCOMMA() {
    let s0;
    if (input.charCodeAt(peg$currPos) === 44) {
      s0 = peg$c36;
      peg$currPos++;
    } else {
      s0 = peg$FAILED;
      if (peg$silentFails === 0) {
        peg$fail(peg$e39);
      }
    }
    return s0;
  }
  function peg$parsePAREN_LEFT() {
    let s0;
    if (input.charCodeAt(peg$currPos) === 40) {
      s0 = peg$c37;
      peg$currPos++;
    } else {
      s0 = peg$FAILED;
      if (peg$silentFails === 0) {
        peg$fail(peg$e40);
      }
    }
    return s0;
  }
  function peg$parsePAREN_RIGHT() {
    let s0;
    if (input.charCodeAt(peg$currPos) === 41) {
      s0 = peg$c38;
      peg$currPos++;
    } else {
      s0 = peg$FAILED;
      if (peg$silentFails === 0) {
        peg$fail(peg$e41);
      }
    }
    return s0;
  }
  function peg$parseQUESTION() {
    let s0;
    if (input.charCodeAt(peg$currPos) === 63) {
      s0 = peg$c39;
      peg$currPos++;
    } else {
      s0 = peg$FAILED;
      if (peg$silentFails === 0) {
        peg$fail(peg$e42);
      }
    }
    return s0;
  }
  function peg$parseREGEX_EXPR() {
    let s0, s1, s2, s3, s4, s5;
    peg$silentFails++;
    s0 = peg$currPos;
    if (input.charCodeAt(peg$currPos) === 47) {
      s1 = peg$c40;
      peg$currPos++;
    } else {
      s1 = peg$FAILED;
      if (peg$silentFails === 0) {
        peg$fail(peg$e44);
      }
    }
    if (s1 !== peg$FAILED) {
      s2 = [];
      s3 = input.charAt(peg$currPos);
      if (peg$r2.test(s3)) {
        peg$currPos++;
      } else {
        s3 = peg$FAILED;
        if (peg$silentFails === 0) {
          peg$fail(peg$e45);
        }
      }
      if (s3 === peg$FAILED) {
        s3 = peg$currPos;
        if (input.charCodeAt(peg$currPos) === 92) {
          s4 = peg$c41;
          peg$currPos++;
        } else {
          s4 = peg$FAILED;
          if (peg$silentFails === 0) {
            peg$fail(peg$e46);
          }
        }
        if (s4 !== peg$FAILED) {
          if (input.length > peg$currPos) {
            s5 = input.charAt(peg$currPos);
            peg$currPos++;
          } else {
            s5 = peg$FAILED;
            if (peg$silentFails === 0) {
              peg$fail(peg$e47);
            }
          }
          if (s5 !== peg$FAILED) {
            s4 = [s4, s5];
            s3 = s4;
          } else {
            peg$currPos = s3;
            s3 = peg$FAILED;
          }
        } else {
          peg$currPos = s3;
          s3 = peg$FAILED;
        }
      }
      while (s3 !== peg$FAILED) {
        s2.push(s3);
        s3 = input.charAt(peg$currPos);
        if (peg$r2.test(s3)) {
          peg$currPos++;
        } else {
          s3 = peg$FAILED;
          if (peg$silentFails === 0) {
            peg$fail(peg$e45);
          }
        }
        if (s3 === peg$FAILED) {
          s3 = peg$currPos;
          if (input.charCodeAt(peg$currPos) === 92) {
            s4 = peg$c41;
            peg$currPos++;
          } else {
            s4 = peg$FAILED;
            if (peg$silentFails === 0) {
              peg$fail(peg$e46);
            }
          }
          if (s4 !== peg$FAILED) {
            if (input.length > peg$currPos) {
              s5 = input.charAt(peg$currPos);
              peg$currPos++;
            } else {
              s5 = peg$FAILED;
              if (peg$silentFails === 0) {
                peg$fail(peg$e47);
              }
            }
            if (s5 !== peg$FAILED) {
              s4 = [s4, s5];
              s3 = s4;
            } else {
              peg$currPos = s3;
              s3 = peg$FAILED;
            }
          } else {
            peg$currPos = s3;
            s3 = peg$FAILED;
          }
        }
      }
      if (input.charCodeAt(peg$currPos) === 47) {
        s3 = peg$c40;
        peg$currPos++;
      } else {
        s3 = peg$FAILED;
        if (peg$silentFails === 0) {
          peg$fail(peg$e44);
        }
      }
      if (s3 !== peg$FAILED) {
        peg$savedPos = s0;
        s0 = peg$f76(s2);
      } else {
        peg$currPos = s0;
        s0 = peg$FAILED;
      }
    } else {
      peg$currPos = s0;
      s0 = peg$FAILED;
    }
    peg$silentFails--;
    if (s0 === peg$FAILED) {
      s1 = peg$FAILED;
      if (peg$silentFails === 0) {
        peg$fail(peg$e43);
      }
    }
    return s0;
  }
  function peg$parseREGEX_OPT() {
    let s0, s1, s2;
    s0 = peg$currPos;
    s1 = [];
    s2 = input.charAt(peg$currPos);
    if (peg$r3.test(s2)) {
      peg$currPos++;
    } else {
      s2 = peg$FAILED;
      if (peg$silentFails === 0) {
        peg$fail(peg$e48);
      }
    }
    while (s2 !== peg$FAILED) {
      s1.push(s2);
      s2 = input.charAt(peg$currPos);
      if (peg$r3.test(s2)) {
        peg$currPos++;
      } else {
        s2 = peg$FAILED;
        if (peg$silentFails === 0) {
          peg$fail(peg$e48);
        }
      }
    }
    peg$savedPos = s0;
    s1 = peg$f77(s1);
    s0 = s1;
    return s0;
  }
  function peg$parseKEY() {
    let s0, s1, s2, s3;
    s0 = peg$currPos;
    s1 = input.charAt(peg$currPos);
    if (peg$r4.test(s1)) {
      peg$currPos++;
    } else {
      s1 = peg$FAILED;
      if (peg$silentFails === 0) {
        peg$fail(peg$e49);
      }
    }
    if (s1 !== peg$FAILED) {
      s2 = [];
      s3 = peg$parsekeyChar();
      while (s3 !== peg$FAILED) {
        s2.push(s3);
        s3 = peg$parsekeyChar();
      }
      peg$savedPos = s0;
      s0 = peg$f78(s1, s2);
    } else {
      peg$currPos = s0;
      s0 = peg$FAILED;
    }
    return s0;
  }
  function peg$parsekeyChar() {
    let s0, s1, s2, s3;
    s0 = input.charAt(peg$currPos);
    if (peg$r5.test(s0)) {
      peg$currPos++;
    } else {
      s0 = peg$FAILED;
      if (peg$silentFails === 0) {
        peg$fail(peg$e50);
      }
    }
    if (s0 === peg$FAILED) {
      s0 = peg$currPos;
      if (input.charCodeAt(peg$currPos) === 45) {
        s1 = peg$c42;
        peg$currPos++;
      } else {
        s1 = peg$FAILED;
        if (peg$silentFails === 0) {
          peg$fail(peg$e51);
        }
      }
      if (s1 !== peg$FAILED) {
        s2 = peg$currPos;
        peg$silentFails++;
        s3 = input.charAt(peg$currPos);
        if (peg$r6.test(s3)) {
          peg$currPos++;
        } else {
          s3 = peg$FAILED;
          if (peg$silentFails === 0) {
            peg$fail(peg$e52);
          }
        }
        peg$silentFails--;
        if (s3 === peg$FAILED) {
          s2 = void 0;
        } else {
          peg$currPos = s2;
          s2 = peg$FAILED;
        }
        if (s2 !== peg$FAILED) {
          peg$savedPos = s0;
          s0 = peg$f79();
        } else {
          peg$currPos = s0;
          s0 = peg$FAILED;
        }
      } else {
        peg$currPos = s0;
        s0 = peg$FAILED;
      }
    }
    return s0;
  }
  function peg$parseSPECIAL_KEY() {
    let s0, s1, s2;
    s0 = peg$currPos;
    s1 = [];
    s2 = input.charAt(peg$currPos);
    if (peg$r7.test(s2)) {
      peg$currPos++;
    } else {
      s2 = peg$FAILED;
      if (peg$silentFails === 0) {
        peg$fail(peg$e53);
      }
    }
    if (s2 !== peg$FAILED) {
      while (s2 !== peg$FAILED) {
        s1.push(s2);
        s2 = input.charAt(peg$currPos);
        if (peg$r7.test(s2)) {
          peg$currPos++;
        } else {
          s2 = peg$FAILED;
          if (peg$silentFails === 0) {
            peg$fail(peg$e53);
          }
        }
      }
    } else {
      s1 = peg$FAILED;
    }
    if (s1 !== peg$FAILED) {
      peg$savedPos = s0;
      s1 = peg$f80(s1);
    }
    s0 = s1;
    return s0;
  }
  function peg$parseNUMBER() {
    let s0, s1, s2, s3, s4, s5, s6;
    peg$silentFails++;
    s0 = peg$currPos;
    if (input.charCodeAt(peg$currPos) === 45) {
      s1 = peg$c42;
      peg$currPos++;
    } else {
      s1 = peg$FAILED;
      if (peg$silentFails === 0) {
        peg$fail(peg$e51);
      }
    }
    if (s1 === peg$FAILED) {
      s1 = null;
    }
    s2 = peg$parseINT();
    if (s2 !== peg$FAILED) {
      s3 = peg$currPos;
      if (input.charCodeAt(peg$currPos) === 46) {
        s4 = peg$c8;
        peg$currPos++;
      } else {
        s4 = peg$FAILED;
        if (peg$silentFails === 0) {
          peg$fail(peg$e11);
        }
      }
      if (s4 !== peg$FAILED) {
        s5 = [];
        s6 = input.charAt(peg$currPos);
        if (peg$r8.test(s6)) {
          peg$currPos++;
        } else {
          s6 = peg$FAILED;
          if (peg$silentFails === 0) {
            peg$fail(peg$e55);
          }
        }
        if (s6 !== peg$FAILED) {
          while (s6 !== peg$FAILED) {
            s5.push(s6);
            s6 = input.charAt(peg$currPos);
            if (peg$r8.test(s6)) {
              peg$currPos++;
            } else {
              s6 = peg$FAILED;
              if (peg$silentFails === 0) {
                peg$fail(peg$e55);
              }
            }
          }
        } else {
          s5 = peg$FAILED;
        }
        if (s5 !== peg$FAILED) {
          s4 = [s4, s5];
          s3 = s4;
        } else {
          peg$currPos = s3;
          s3 = peg$FAILED;
        }
      } else {
        peg$currPos = s3;
        s3 = peg$FAILED;
      }
      if (s3 === peg$FAILED) {
        s3 = null;
      }
      s4 = peg$parseEXP();
      if (s4 === peg$FAILED) {
        s4 = null;
      }
      peg$savedPos = s0;
      s0 = peg$f81(s3, s4);
    } else {
      peg$currPos = s0;
      s0 = peg$FAILED;
    }
    peg$silentFails--;
    if (s0 === peg$FAILED) {
      s1 = peg$FAILED;
      if (peg$silentFails === 0) {
        peg$fail(peg$e54);
      }
    }
    return s0;
  }
  function peg$parseINT() {
    let s0, s1;
    s0 = [];
    s1 = input.charAt(peg$currPos);
    if (peg$r8.test(s1)) {
      peg$currPos++;
    } else {
      s1 = peg$FAILED;
      if (peg$silentFails === 0) {
        peg$fail(peg$e55);
      }
    }
    if (s1 !== peg$FAILED) {
      while (s1 !== peg$FAILED) {
        s0.push(s1);
        s1 = input.charAt(peg$currPos);
        if (peg$r8.test(s1)) {
          peg$currPos++;
        } else {
          s1 = peg$FAILED;
          if (peg$silentFails === 0) {
            peg$fail(peg$e55);
          }
        }
      }
    } else {
      s0 = peg$FAILED;
    }
    return s0;
  }
  function peg$parseEXP() {
    let s0, s1, s2, s3, s4;
    s0 = peg$currPos;
    s1 = input.charAt(peg$currPos);
    if (peg$r9.test(s1)) {
      peg$currPos++;
    } else {
      s1 = peg$FAILED;
      if (peg$silentFails === 0) {
        peg$fail(peg$e56);
      }
    }
    if (s1 !== peg$FAILED) {
      s2 = input.charAt(peg$currPos);
      if (peg$r0.test(s2)) {
        peg$currPos++;
      } else {
        s2 = peg$FAILED;
        if (peg$silentFails === 0) {
          peg$fail(peg$e5);
        }
      }
      if (s2 === peg$FAILED) {
        s2 = null;
      }
      s3 = [];
      s4 = input.charAt(peg$currPos);
      if (peg$r8.test(s4)) {
        peg$currPos++;
      } else {
        s4 = peg$FAILED;
        if (peg$silentFails === 0) {
          peg$fail(peg$e55);
        }
      }
      if (s4 !== peg$FAILED) {
        while (s4 !== peg$FAILED) {
          s3.push(s4);
          s4 = input.charAt(peg$currPos);
          if (peg$r8.test(s4)) {
            peg$currPos++;
          } else {
            s4 = peg$FAILED;
            if (peg$silentFails === 0) {
              peg$fail(peg$e55);
            }
          }
        }
      } else {
        s3 = peg$FAILED;
      }
      if (s3 !== peg$FAILED) {
        s1 = [s1, s2, s3];
        s0 = s1;
      } else {
        peg$currPos = s0;
        s0 = peg$FAILED;
      }
    } else {
      peg$currPos = s0;
      s0 = peg$FAILED;
    }
    return s0;
  }
  function peg$parseSTRING() {
    let s0, s1, s2, s3;
    peg$silentFails++;
    s0 = peg$currPos;
    if (input.charCodeAt(peg$currPos) === 34) {
      s1 = peg$c43;
      peg$currPos++;
    } else {
      s1 = peg$FAILED;
      if (peg$silentFails === 0) {
        peg$fail(peg$e57);
      }
    }
    if (s1 !== peg$FAILED) {
      s2 = [];
      s3 = peg$parsedoubleStringChar();
      while (s3 !== peg$FAILED) {
        s2.push(s3);
        s3 = peg$parsedoubleStringChar();
      }
      if (input.charCodeAt(peg$currPos) === 34) {
        s3 = peg$c43;
        peg$currPos++;
      } else {
        s3 = peg$FAILED;
        if (peg$silentFails === 0) {
          peg$fail(peg$e57);
        }
      }
      if (s3 !== peg$FAILED) {
        peg$savedPos = s0;
        s0 = peg$f82(s2);
      } else {
        peg$currPos = s0;
        s0 = peg$FAILED;
      }
    } else {
      peg$currPos = s0;
      s0 = peg$FAILED;
    }
    if (s0 === peg$FAILED) {
      s0 = peg$currPos;
      if (input.charCodeAt(peg$currPos) === 39) {
        s1 = peg$c44;
        peg$currPos++;
      } else {
        s1 = peg$FAILED;
        if (peg$silentFails === 0) {
          peg$fail(peg$e58);
        }
      }
      if (s1 !== peg$FAILED) {
        s2 = [];
        s3 = peg$parsesingleStringChar();
        while (s3 !== peg$FAILED) {
          s2.push(s3);
          s3 = peg$parsesingleStringChar();
        }
        if (input.charCodeAt(peg$currPos) === 39) {
          s3 = peg$c44;
          peg$currPos++;
        } else {
          s3 = peg$FAILED;
          if (peg$silentFails === 0) {
            peg$fail(peg$e58);
          }
        }
        if (s3 !== peg$FAILED) {
          peg$savedPos = s0;
          s0 = peg$f83(s2);
        } else {
          peg$currPos = s0;
          s0 = peg$FAILED;
        }
      } else {
        peg$currPos = s0;
        s0 = peg$FAILED;
      }
    }
    peg$silentFails--;
    if (s0 === peg$FAILED) {
      s1 = peg$FAILED;
      if (peg$silentFails === 0) {
        peg$fail(peg$e7);
      }
    }
    return s0;
  }
  function peg$parsedoubleStringChar() {
    let s0, s1, s2;
    s0 = peg$currPos;
    s1 = peg$currPos;
    peg$silentFails++;
    s2 = input.charAt(peg$currPos);
    if (peg$r10.test(s2)) {
      peg$currPos++;
    } else {
      s2 = peg$FAILED;
      if (peg$silentFails === 0) {
        peg$fail(peg$e59);
      }
    }
    peg$silentFails--;
    if (s2 === peg$FAILED) {
      s1 = void 0;
    } else {
      peg$currPos = s1;
      s1 = peg$FAILED;
    }
    if (s1 !== peg$FAILED) {
      if (input.length > peg$currPos) {
        s2 = input.charAt(peg$currPos);
        peg$currPos++;
      } else {
        s2 = peg$FAILED;
        if (peg$silentFails === 0) {
          peg$fail(peg$e47);
        }
      }
      if (s2 !== peg$FAILED) {
        peg$savedPos = s0;
        s0 = peg$f84(s2);
      } else {
        peg$currPos = s0;
        s0 = peg$FAILED;
      }
    } else {
      peg$currPos = s0;
      s0 = peg$FAILED;
    }
    if (s0 === peg$FAILED) {
      s0 = peg$currPos;
      if (input.charCodeAt(peg$currPos) === 92) {
        s1 = peg$c41;
        peg$currPos++;
      } else {
        s1 = peg$FAILED;
        if (peg$silentFails === 0) {
          peg$fail(peg$e46);
        }
      }
      if (s1 !== peg$FAILED) {
        s2 = peg$parseescapeSequence();
        if (s2 !== peg$FAILED) {
          peg$savedPos = s0;
          s0 = peg$f85(s2);
        } else {
          peg$currPos = s0;
          s0 = peg$FAILED;
        }
      } else {
        peg$currPos = s0;
        s0 = peg$FAILED;
      }
    }
    return s0;
  }
  function peg$parsesingleStringChar() {
    let s0, s1, s2;
    s0 = peg$currPos;
    s1 = peg$currPos;
    peg$silentFails++;
    s2 = input.charAt(peg$currPos);
    if (peg$r11.test(s2)) {
      peg$currPos++;
    } else {
      s2 = peg$FAILED;
      if (peg$silentFails === 0) {
        peg$fail(peg$e60);
      }
    }
    peg$silentFails--;
    if (s2 === peg$FAILED) {
      s1 = void 0;
    } else {
      peg$currPos = s1;
      s1 = peg$FAILED;
    }
    if (s1 !== peg$FAILED) {
      if (input.length > peg$currPos) {
        s2 = input.charAt(peg$currPos);
        peg$currPos++;
      } else {
        s2 = peg$FAILED;
        if (peg$silentFails === 0) {
          peg$fail(peg$e47);
        }
      }
      if (s2 !== peg$FAILED) {
        peg$savedPos = s0;
        s0 = peg$f86(s2);
      } else {
        peg$currPos = s0;
        s0 = peg$FAILED;
      }
    } else {
      peg$currPos = s0;
      s0 = peg$FAILED;
    }
    if (s0 === peg$FAILED) {
      s0 = peg$currPos;
      if (input.charCodeAt(peg$currPos) === 92) {
        s1 = peg$c41;
        peg$currPos++;
      } else {
        s1 = peg$FAILED;
        if (peg$silentFails === 0) {
          peg$fail(peg$e46);
        }
      }
      if (s1 !== peg$FAILED) {
        s2 = peg$parseescapeSequence();
        if (s2 !== peg$FAILED) {
          peg$savedPos = s0;
          s0 = peg$f87(s2);
        } else {
          peg$currPos = s0;
          s0 = peg$FAILED;
        }
      } else {
        peg$currPos = s0;
        s0 = peg$FAILED;
      }
    }
    return s0;
  }
  function peg$parseescapeSequence() {
    let s0, s1, s2, s3, s4, s5, s6, s7;
    s0 = peg$currPos;
    if (input.charCodeAt(peg$currPos) === 34) {
      s1 = peg$c43;
      peg$currPos++;
    } else {
      s1 = peg$FAILED;
      if (peg$silentFails === 0) {
        peg$fail(peg$e57);
      }
    }
    if (s1 !== peg$FAILED) {
      peg$savedPos = s0;
      s1 = peg$f88();
    }
    s0 = s1;
    if (s0 === peg$FAILED) {
      s0 = peg$currPos;
      if (input.charCodeAt(peg$currPos) === 39) {
        s1 = peg$c44;
        peg$currPos++;
      } else {
        s1 = peg$FAILED;
        if (peg$silentFails === 0) {
          peg$fail(peg$e58);
        }
      }
      if (s1 !== peg$FAILED) {
        peg$savedPos = s0;
        s1 = peg$f89();
      }
      s0 = s1;
      if (s0 === peg$FAILED) {
        s0 = peg$currPos;
        if (input.charCodeAt(peg$currPos) === 92) {
          s1 = peg$c41;
          peg$currPos++;
        } else {
          s1 = peg$FAILED;
          if (peg$silentFails === 0) {
            peg$fail(peg$e46);
          }
        }
        if (s1 !== peg$FAILED) {
          peg$savedPos = s0;
          s1 = peg$f90();
        }
        s0 = s1;
        if (s0 === peg$FAILED) {
          s0 = peg$currPos;
          if (input.charCodeAt(peg$currPos) === 47) {
            s1 = peg$c40;
            peg$currPos++;
          } else {
            s1 = peg$FAILED;
            if (peg$silentFails === 0) {
              peg$fail(peg$e44);
            }
          }
          if (s1 !== peg$FAILED) {
            peg$savedPos = s0;
            s1 = peg$f91();
          }
          s0 = s1;
          if (s0 === peg$FAILED) {
            s0 = peg$currPos;
            if (input.charCodeAt(peg$currPos) === 98) {
              s1 = peg$c45;
              peg$currPos++;
            } else {
              s1 = peg$FAILED;
              if (peg$silentFails === 0) {
                peg$fail(peg$e61);
              }
            }
            if (s1 !== peg$FAILED) {
              peg$savedPos = s0;
              s1 = peg$f92();
            }
            s0 = s1;
            if (s0 === peg$FAILED) {
              s0 = peg$currPos;
              if (input.charCodeAt(peg$currPos) === 102) {
                s1 = peg$c46;
                peg$currPos++;
              } else {
                s1 = peg$FAILED;
                if (peg$silentFails === 0) {
                  peg$fail(peg$e62);
                }
              }
              if (s1 !== peg$FAILED) {
                peg$savedPos = s0;
                s1 = peg$f93();
              }
              s0 = s1;
              if (s0 === peg$FAILED) {
                s0 = peg$currPos;
                if (input.charCodeAt(peg$currPos) === 110) {
                  s1 = peg$c47;
                  peg$currPos++;
                } else {
                  s1 = peg$FAILED;
                  if (peg$silentFails === 0) {
                    peg$fail(peg$e63);
                  }
                }
                if (s1 !== peg$FAILED) {
                  peg$savedPos = s0;
                  s1 = peg$f94();
                }
                s0 = s1;
                if (s0 === peg$FAILED) {
                  s0 = peg$currPos;
                  if (input.charCodeAt(peg$currPos) === 114) {
                    s1 = peg$c48;
                    peg$currPos++;
                  } else {
                    s1 = peg$FAILED;
                    if (peg$silentFails === 0) {
                      peg$fail(peg$e64);
                    }
                  }
                  if (s1 !== peg$FAILED) {
                    peg$savedPos = s0;
                    s1 = peg$f95();
                  }
                  s0 = s1;
                  if (s0 === peg$FAILED) {
                    s0 = peg$currPos;
                    if (input.charCodeAt(peg$currPos) === 116) {
                      s1 = peg$c49;
                      peg$currPos++;
                    } else {
                      s1 = peg$FAILED;
                      if (peg$silentFails === 0) {
                        peg$fail(peg$e65);
                      }
                    }
                    if (s1 !== peg$FAILED) {
                      peg$savedPos = s0;
                      s1 = peg$f96();
                    }
                    s0 = s1;
                    if (s0 === peg$FAILED) {
                      s0 = peg$currPos;
                      if (input.charCodeAt(peg$currPos) === 117) {
                        s1 = peg$c50;
                        peg$currPos++;
                      } else {
                        s1 = peg$FAILED;
                        if (peg$silentFails === 0) {
                          peg$fail(peg$e66);
                        }
                      }
                      if (s1 !== peg$FAILED) {
                        s2 = peg$currPos;
                        s3 = peg$currPos;
                        s4 = peg$parseHEX();
                        if (s4 !== peg$FAILED) {
                          s5 = peg$parseHEX();
                          if (s5 !== peg$FAILED) {
                            s6 = peg$parseHEX();
                            if (s6 !== peg$FAILED) {
                              s7 = peg$parseHEX();
                              if (s7 !== peg$FAILED) {
                                s4 = [s4, s5, s6, s7];
                                s3 = s4;
                              } else {
                                peg$currPos = s3;
                                s3 = peg$FAILED;
                              }
                            } else {
                              peg$currPos = s3;
                              s3 = peg$FAILED;
                            }
                          } else {
                            peg$currPos = s3;
                            s3 = peg$FAILED;
                          }
                        } else {
                          peg$currPos = s3;
                          s3 = peg$FAILED;
                        }
                        if (s3 !== peg$FAILED) {
                          s2 = input.substring(s2, peg$currPos);
                        } else {
                          s2 = s3;
                        }
                        if (s2 !== peg$FAILED) {
                          peg$savedPos = s0;
                          s0 = peg$f97(s2);
                        } else {
                          peg$currPos = s0;
                          s0 = peg$FAILED;
                        }
                      } else {
                        peg$currPos = s0;
                        s0 = peg$FAILED;
                      }
                    }
                  }
                }
              }
            }
          }
        }
      }
    }
    return s0;
  }
  function peg$parseHEX() {
    let s0;
    s0 = input.charAt(peg$currPos);
    if (peg$r12.test(s0)) {
      peg$currPos++;
    } else {
      s0 = peg$FAILED;
      if (peg$silentFails === 0) {
        peg$fail(peg$e67);
      }
    }
    return s0;
  }
  function peg$parseidentifierChar() {
    let s0;
    s0 = input.charAt(peg$currPos);
    if (peg$r5.test(s0)) {
      peg$currPos++;
    } else {
      s0 = peg$FAILED;
      if (peg$silentFails === 0) {
        peg$fail(peg$e50);
      }
    }
    return s0;
  }
  function peg$parse_() {
    let s0, s1;
    peg$silentFails++;
    s0 = [];
    s1 = input.charAt(peg$currPos);
    if (peg$r13.test(s1)) {
      peg$currPos++;
    } else {
      s1 = peg$FAILED;
      if (peg$silentFails === 0) {
        peg$fail(peg$e68);
      }
    }
    while (s1 !== peg$FAILED) {
      s0.push(s1);
      s1 = input.charAt(peg$currPos);
      if (peg$r13.test(s1)) {
        peg$currPos++;
      } else {
        s1 = peg$FAILED;
        if (peg$silentFails === 0) {
          peg$fail(peg$e68);
        }
      }
    }
    peg$silentFails--;
    return s0;
  }
  function peg$parseEOF() {
    let s0, s1;
    s0 = peg$currPos;
    peg$silentFails++;
    if (input.length > peg$currPos) {
      s1 = input.charAt(peg$currPos);
      peg$currPos++;
    } else {
      s1 = peg$FAILED;
      if (peg$silentFails === 0) {
        peg$fail(peg$e47);
      }
    }
    peg$silentFails--;
    if (s1 === peg$FAILED) {
      s0 = void 0;
    } else {
      peg$currPos = s0;
      s0 = peg$FAILED;
    }
    return s0;
  }
  peg$result = peg$startRuleFunction();
  const peg$success = peg$result !== peg$FAILED && peg$currPos === input.length;
  function peg$throw() {
    if (peg$result !== peg$FAILED && peg$currPos < input.length) {
      peg$fail(peg$endExpectation());
    }
    throw peg$buildStructuredError(
      peg$maxFailExpected,
      peg$maxFailPos < input.length ? peg$getUnicode(peg$maxFailPos) : null,
      peg$maxFailPos < input.length ? peg$computeLocation(peg$maxFailPos, peg$maxFailPos + 1) : peg$computeLocation(peg$maxFailPos, peg$maxFailPos)
    );
  }
  if (options.peg$library) {
    return (
      /** @type {any} */
      {
        peg$result,
        peg$currPos,
        peg$FAILED,
        peg$maxFailExpected,
        peg$maxFailPos,
        peg$success,
        peg$throw: peg$success ? void 0 : peg$throw
      }
    );
  }
  if (peg$success) {
    return peg$result;
  } else {
    peg$throw();
  }
}

// src/parser/parse.ts
function parseInternal(input) {
  try {
    return peg$parse(input);
  } catch (e) {
    if (e && typeof e === "object" && "location" in e && "message" in e) {
      const peggyError = e;
      throw new JSONPathSyntaxError(
        peggyError.location.start.line,
        peggyError.location.start.column,
        peggyError.message
      );
    }
    throw e;
  }
}
function parse(input, options = {}) {
  try {
    return parseInternal(input);
  } catch (e) {
    if (!options.hideExceptions) {
      throw e;
    }
    return null;
  }
}

// src/handler/helper.ts
var isEqual = (a, b) => {
  if (a === b) return true;
  if (a === null || b === null || a === void 0 || b === void 0) {
    return a === b;
  }
  if (typeof a !== typeof b) return false;
  if (Array.isArray(a) && Array.isArray(b)) {
    if (a.length !== b.length) return false;
    for (let i = 0; i < a.length; i++) {
      if (!isEqual(a[i], b[i])) return false;
    }
    return true;
  }
  if (Array.isArray(a) !== Array.isArray(b)) return false;
  if (typeof a === "object" && typeof b === "object") {
    const keysA = Object.keys(a);
    const keysB = Object.keys(b);
    if (keysA.length !== keysB.length) return false;
    for (const key of keysA) {
      if (!Object.prototype.hasOwnProperty.call(b, key)) return false;
      if (!isEqual(a[key], b[key])) {
        return false;
      }
    }
    return true;
  }
  return false;
};
var isArray = (item) => {
  return Array.isArray(item);
};
var isPlainObject = (item) => {
  if (typeof item !== "object" || item === null) {
    return false;
  }
  let proto = item;
  while (Object.getPrototypeOf(proto) !== null) {
    proto = Object.getPrototypeOf(proto);
  }
  return Object.getPrototypeOf(item) === proto;
};
var isNumber = (item) => {
  return typeof item === "number";
};
var isString = (item) => {
  return typeof item === "string";
};
var isUndefined = (item) => {
  return typeof item === "undefined";
};
var isDefined = (item) => {
  return typeof item !== "undefined";
};

// src/handler/functions.ts
var hasWordBoundaryOutsideCharClass = (pattern) => {
  let inCharClass = false;
  for (let i = 0; i < pattern.length; i++) {
    const char = pattern[i];
    if (char === "\\" && i + 1 < pattern.length) {
      const next = pattern[i + 1];
      if ((next === "b" || next === "B") && !inCharClass) {
        return true;
      }
      i++;
    } else if (char === "[" && !inCharClass) {
      inCharClass = true;
    } else if (char === "]" && inCharClass) {
      inCharClass = false;
    }
  }
  return false;
};
var isValidIRegexp = (pattern) => {
  if (/\\[1-9]/.test(pattern)) return false;
  if (/\(\?[=!]/.test(pattern)) return false;
  if (/\(\?<[=!]/.test(pattern)) return false;
  if (/\(\?<[a-zA-Z]/.test(pattern)) return false;
  if (hasWordBoundaryOutsideCharClass(pattern)) return false;
  return true;
};
var length = (args) => {
  const val = args[0];
  if (isString(val)) return val.length;
  if (isArray(val)) return val.length;
  if (isPlainObject(val)) return Object.keys(val).length;
  return void 0;
};
var count = (args) => {
  const val = args[0];
  if (isArray(val)) return val.length;
  return isDefined(val) ? 1 : 0;
};
var match = (args) => {
  const val = args[0];
  const pattern = args[1];
  if (!isString(val) || !isString(pattern)) return false;
  if (!isValidIRegexp(pattern)) return false;
  try {
    const regex = new RegExp(`^(?:${pattern})$`, "u");
    return regex.test(val);
  } catch (e) {
    return false;
  }
};
var search = (args) => {
  const val = args[0];
  const pattern = args[1];
  if (!isString(val) || !isString(pattern)) return false;
  if (!isValidIRegexp(pattern)) return false;
  try {
    const regex = new RegExp(pattern, "u");
    return regex.test(val);
  } catch (e) {
    return false;
  }
};
var value = (args) => {
  const val = args[0];
  if (isArray(val) && val.length === 1) return val[0];
  if (!isArray(val) && isDefined(val)) return val;
  return void 0;
};
var rfc9535Functions = {
  length,
  count,
  match,
  search,
  value
};

// src/handler/comparators.ts
var eq = (left, right) => {
  if (left === void 0 && right === void 0) return true;
  return isEqual(left, right);
};
var ne = (left, right) => !isEqual(left, right);
var lt = (left, right) => {
  if (isNumber(left) && isNumber(right)) return left < right;
  if (isString(left) && isString(right)) return left < right;
  return false;
};
var le = (left, right) => {
  if (isNumber(left) && isNumber(right)) return left <= right;
  if (isString(left) && isString(right)) return left <= right;
  return false;
};
var gt = (left, right) => {
  if (isNumber(left) && isNumber(right)) return left > right;
  if (isString(left) && isString(right)) return left > right;
  return false;
};
var ge = (left, right) => {
  if (isNumber(left) && isNumber(right)) return left >= right;
  if (isString(left) && isString(right)) return left >= right;
  return false;
};
var inOp = (left, right) => {
  if (!isArray(right)) return false;
  return right.includes(left);
};
var nin = (left, right) => {
  if (!isArray(right)) return false;
  return !right.includes(left);
};
var subsetof = (left, right) => {
  if (!isArray(left) || !isArray(right)) return false;
  const rightSet = new Set(right);
  return left.every((e) => rightSet.has(e));
};
var anyof = (left, right) => {
  if (!isArray(left) || !isArray(right)) return false;
  const rightSet = new Set(right);
  return left.some((e) => rightSet.has(e));
};
var noneof = (left, right) => {
  if (!isArray(left) || !isArray(right)) return false;
  const rightSet = new Set(right);
  return !left.some((e) => rightSet.has(e));
};
var sizeof = (left, right) => {
  if (!isArray(left) && !isString(left) || !isArray(right) && !isString(right)) {
    return false;
  }
  return left.length === right.length;
};
var size = (left, right) => {
  if (!isArray(left) && !isString(left) || !isNumber(right)) {
    return false;
  }
  return left.length === right;
};
var comparators = {
  eq,
  ne,
  lt,
  le,
  gt,
  ge,
  in: inOp,
  nin,
  subsetof,
  anyof,
  noneof,
  sizeof,
  size
};
var isEmpty = (value2) => {
  if (!isArray(value2) && !isString(value2)) return false;
  return value2.length === 0;
};
var matchRegex = (left, pattern, opts) => {
  if (!isString(left) || !isString(pattern)) return false;
  const value2 = pattern.slice(1, -1);
  if (!isValidIRegexp(value2)) return false;
  try {
    return !!left.match(new RegExp(value2, opts));
  } catch (e) {
    return false;
  }
};

// src/handler/Handler.ts
var formatStringLiteralPath = (paths2, v) => {
  const escaped = v.replace(/\\/g, "\\\\").replace(/'/g, "\\'").replace(/\x08/g, "\\b").replace(/\f/g, "\\f").replace(/\n/g, "\\n").replace(/\r/g, "\\r").replace(/\t/g, "\\t").replace(/[\x00-\x07\x0b\x0e-\x1f]/g, (c) => "\\u00" + c.charCodeAt(0).toString(16).padStart(2, "0"));
  return paths2.concat(`['${escaped}']`);
};
var formatNumericLiteralPath = (paths2, v) => paths2.concat(`[${v}]`);
var normalizeArrayIndex = (index, length2) => index < 0 ? length2 + index : index;
var getArrayElement = (arr, index) => {
  const normalized = normalizeArrayIndex(index, arr.length);
  if (normalized >= 0 && normalized < arr.length) {
    return { exists: true, value: arr[normalized], index: normalized };
  }
  return { exists: false };
};
var Handler = class {
  constructor(rootPayload) {
    // ============================================
    // Property Access Handlers
    // ============================================
    this.handleIdentifier = (payload, tree) => {
      if (!isPlainObject(payload.value) || !(tree.value in payload.value)) {
        return;
      }
      return { value: payload.value[tree.value], paths: formatStringLiteralPath(payload.paths, tree.value) };
    };
    this.handleStringLiteral = ({ value: value2, paths: paths2 }, tree) => {
      if (!isPlainObject(value2) || !(tree.value in value2)) {
        return;
      }
      return { value: value2[tree.value], paths: formatStringLiteralPath(paths2, tree.value) };
    };
    this.handleWildcard = ({ value: value2, paths: paths2 }) => {
      if (isArray(value2)) {
        return value2.map((item, index) => ({
          value: item,
          paths: formatNumericLiteralPath(paths2, index)
        }));
      }
      if (isPlainObject(value2)) {
        return Object.keys(value2).map((key) => ({
          value: value2[key],
          paths: formatStringLiteralPath(paths2, key)
        }));
      }
      return [];
    };
    // ============================================
    // Numeric Index Handlers
    // ============================================
    /** Bracket notation numeric index - RFC 9535: only works on arrays */
    this.handleNumericLiteral = ({ value: value2, paths: paths2 }, tree) => {
      if (!isArray(value2)) {
        return;
      }
      const result = getArrayElement(value2, tree.value);
      if (result.exists) {
        return { value: result.value, paths: formatNumericLiteralPath(paths2, result.index) };
      }
      return;
    };
    /** Dot notation numeric ($.2) - works as property access on objects, index on arrays */
    this.handleDotNumericLiteral = ({ value: value2, paths: paths2 }, tree) => {
      if (isArray(value2)) {
        const result = getArrayElement(value2, tree.value);
        if (result.exists) {
          return { value: result.value, paths: formatNumericLiteralPath(paths2, result.index) };
        }
        return;
      }
      if (isPlainObject(value2)) {
        const key = String(tree.value);
        if (key in value2) {
          return { value: value2[key], paths: formatStringLiteralPath(paths2, key) };
        }
        return;
      }
      return;
    };
    // ============================================
    // Slice Handler
    // ============================================
    /** RFC 9535 Section 2.3.4.2: Array slice semantics */
    this.handleSlices = ({ value: value2, paths: paths2 }, tree) => {
      var _a;
      if (!isArray(value2)) {
        return [];
      }
      const len = value2.length;
      const step = (_a = tree.step) != null ? _a : 1;
      if (step === 0) {
        return [];
      }
      let n;
      let m;
      if (step > 0) {
        n = tree.start !== null ? normalizeArrayIndex(tree.start, len) : 0;
        m = tree.end !== null ? normalizeArrayIndex(tree.end, len) : len;
      } else {
        n = tree.start !== null ? normalizeArrayIndex(tree.start, len) : len - 1;
        m = tree.end !== null ? normalizeArrayIndex(tree.end, len) : -len - 1;
      }
      let lower;
      let upper;
      if (step > 0) {
        lower = Math.max(n, 0);
        upper = Math.min(m, len);
      } else {
        upper = Math.min(n, len - 1);
        lower = Math.max(m, -1);
      }
      const results = [];
      let i = step > 0 ? lower : upper;
      while (step > 0 && i < upper || step < 0 && i > lower) {
        results.push({ value: value2[i], paths: formatNumericLiteralPath(paths2, i) });
        i += step;
      }
      return results;
    };
    // ============================================
    // Union and Index Handlers
    // ============================================
    // RFC 9535: Union selector list supports mixed types (strings, numbers, wildcards)
    this.handleUnions = (payload, tree) => {
      const results = [];
      for (const selector of tree.values) {
        switch (selector.type) {
          case "identifier":
            if (isPlainObject(payload.value)) {
              const result = this.handleIdentifier(payload, selector);
              if (isDefined(result)) results.push(result);
            }
            break;
          case "stringLiteral":
            if (isPlainObject(payload.value)) {
              const result = this.handleStringLiteral(payload, selector);
              if (isDefined(result)) results.push(result);
            }
            break;
          case "numericLiteral":
            if (isArray(payload.value)) {
              const result = this.handleNumericLiteral(payload, selector);
              if (isDefined(result)) results.push(result);
            }
            break;
          case "wildcard":
            results.push(...this.handleWildcard(payload));
            break;
        }
      }
      return results;
    };
    // ============================================
    // Recursive Descent Handler
    // ============================================
    this.handleDotdot = (payload, tree) => {
      const treeValue = tree.value;
      const value2 = payload.value;
      let results = [];
      switch (treeValue.type) {
        case "bracketMember": {
          const result = this.handleBracketMemberContent(payload, treeValue.value);
          if (isDefined(result)) {
            results = results.concat(result);
          }
          break;
        }
        case "bracketExpression": {
          if (isPlainObject(value2) || isArray(value2) && treeValue.value.type === "wildcard") {
            const result = this.handleBracketExpressionContent(payload, treeValue.value);
            results = results.concat(result);
          }
          break;
        }
        case "wildcard": {
          results = results.concat(this.handleWildcard(payload));
          break;
        }
        case "identifier": {
          const result = this.handleIdentifier(payload, treeValue);
          if (isDefined(result)) {
            results = results.concat(result);
          }
          break;
        }
      }
      if (isPlainObject(value2)) {
        for (const key of Object.keys(value2)) {
          const childPayload = { value: value2[key], paths: formatStringLiteralPath(payload.paths, key) };
          results = results.concat(this.handleDotdot(childPayload, tree));
        }
      } else if (isArray(value2)) {
        value2.forEach((item, index) => {
          const childPayload = { value: item, paths: formatNumericLiteralPath(payload.paths, index) };
          results = results.concat(this.handleDotdot(childPayload, tree));
        });
      }
      return results;
    };
    // ============================================
    // Operation Handlers (Arithmetic)
    // ============================================
    this.handleOperationContent = (payload, tree) => {
      var _a, _b;
      switch (tree.type) {
        case "root":
          return this.handleSubscript(this.rootPayload, tree.next);
        case "current":
          return this.handleSubscript(payload, tree.next);
        case "value":
          return { value: tree.value, paths: payload.paths };
        case "groupOperation":
          return this.handleOperationContent(payload, tree.value);
        case "functionCall": {
          const result = this.handleFunctionCall(payload, tree);
          return { value: result, paths: payload.paths };
        }
        case "operation": {
          const left = (_a = this.handleOperationContent(payload, tree.left)) == null ? void 0 : _a.value;
          const right = (_b = this.handleOperationContent(payload, tree.right)) == null ? void 0 : _b.value;
          if (!isNumber(left) || !isNumber(right)) {
            return;
          }
          switch (tree.operator) {
            case "plus":
              return { value: left + right, paths: payload.paths };
            case "minus":
              return { value: left - right, paths: payload.paths };
            case "div":
              return right === 0 ? void 0 : { value: left / right, paths: payload.paths };
            case "multi":
              return { value: left * right, paths: payload.paths };
          }
        }
      }
    };
    // ============================================
    // Comparator Handler
    // ============================================
    this.handleComparator = (payload, tree) => {
      var _a, _b;
      const leftValue = (_a = this.handleOperationContent(payload, tree.left)) == null ? void 0 : _a.value;
      if (tree.operator === "empty") {
        return isEmpty(leftValue);
      }
      if (tree.operator === "reg") {
        return matchRegex(leftValue, tree.right.value, tree.right.opts);
      }
      const rightValue = (_b = this.handleOperationContent(payload, tree.right)) == null ? void 0 : _b.value;
      const comparatorFn = comparators[tree.operator];
      if (comparatorFn) {
        return comparatorFn(leftValue, rightValue);
      }
      return false;
    };
    // ============================================
    // Logical Expression Handler
    // ============================================
    this.handleLogicalExpression = (payload, tree) => {
      const leftValue = this.handleFilterExpressionContent(payload, tree.left);
      const rightValue = this.handleFilterExpressionContent(payload, tree.right);
      switch (tree.operator) {
        case "and":
          return leftValue && rightValue;
        case "or":
          return leftValue || rightValue;
      }
    };
    // ============================================
    // Function Call Handler
    // ============================================
    this.handleFunctionCall = (payload, tree) => {
      const args = tree.args.map((arg) => {
        var _a;
        return (_a = this.handleOperationContent(payload, arg)) == null ? void 0 : _a.value;
      });
      const fn = rfc9535Functions[tree.name];
      return fn ? fn(args) : void 0;
    };
    // ============================================
    // Filter Expression Handler
    // ============================================
    this.handleFilterExpressionContent = (payload, tree) => {
      switch (tree.type) {
        case "logicalExpression":
          return this.handleLogicalExpression(payload, tree);
        case "comparator":
          return this.handleComparator(payload, tree);
        case "groupExpression":
          return this.handleFilterExpressionContent(payload, tree.value);
        case "notExpression":
          return !this.handleFilterExpressionContent(payload, tree.value);
        case "functionCall":
          return !!this.handleFunctionCall(payload, tree);
        case "root": {
          const result = this.handleSubscript(this.rootPayload, tree.next);
          if (result == null ? void 0 : result.isIndefinite) {
            return Array.isArray(result.value) && result.value.length > 0;
          }
          return isDefined(result);
        }
        case "current": {
          const result = this.handleSubscript(payload, tree.next);
          if (result == null ? void 0 : result.isIndefinite) {
            return Array.isArray(result.value) && result.value.length > 0;
          }
          return isDefined(result);
        }
        case "value":
          return !!tree.value;
      }
    };
    // ============================================
    // Bracket Content Handlers
    // ============================================
    this.handleBracketMemberContent = (payload, tree) => {
      switch (tree.type) {
        case "identifier":
          return this.handleIdentifier(payload, tree);
        case "numericLiteral":
          return this.handleNumericLiteral(payload, tree);
        case "stringLiteral":
          return this.handleStringLiteral(payload, tree);
      }
    };
    this.handleBracketExpressionContent = (payload, tree) => {
      const payloadValue = payload.value;
      switch (tree.type) {
        case "filterExpression": {
          const results = [];
          if (isArray(payloadValue)) {
            payloadValue.forEach((value2, index) => {
              const item = { value: value2, paths: formatNumericLiteralPath(payload.paths, index) };
              if (this.handleFilterExpressionContent(item, tree.value)) {
                results.push(item);
              }
            });
          } else if (isPlainObject(payloadValue)) {
            for (const key of Object.keys(payloadValue)) {
              const item = { value: payloadValue[key], paths: formatStringLiteralPath(payload.paths, key) };
              if (this.handleFilterExpressionContent(item, tree.value)) {
                results.push(item);
              }
            }
          }
          return results;
        }
        case "unions":
          return this.handleUnions(payload, tree);
        case "wildcard":
          return this.handleWildcard(payload);
        case "slices":
          return this.handleSlices(payload, tree);
      }
    };
    // ============================================
    // Result Aggregation
    // ============================================
    this.concatIndefiniteValuePaths = (payload, tree) => {
      if (!tree) {
        return payload.reduce(
          (acc, current) => ({
            isIndefinite: true,
            value: [...acc.value, current.value],
            paths: [...acc.paths, current.paths]
          }),
          { value: [], paths: [], isIndefinite: true }
        );
      }
      let values = [];
      let paths2 = [];
      for (const item of payload) {
        const result = this.handleSubscript(item, tree);
        if (isUndefined(result)) {
          continue;
        }
        values = result.isIndefinite ? values.concat(result.value) : [...values, result.value];
        paths2 = paths2.concat(result.paths);
      }
      return { value: values, paths: paths2, isIndefinite: true };
    };
    // ============================================
    // Main Subscript Handler
    // ============================================
    this.handleSubscript = (payload, tree) => {
      if (tree === null) {
        return payload;
      }
      const treeValue = tree.value;
      switch (treeValue.type) {
        case "bracketExpression": {
          const result = this.handleBracketExpressionContent(payload, treeValue.value);
          return this.concatIndefiniteValuePaths(result, tree.next);
        }
        case "bracketMember": {
          const result = this.handleBracketMemberContent(payload, treeValue.value);
          if (isUndefined(result)) {
            return;
          }
          return this.handleSubscript(result, tree.next);
        }
        case "dot": {
          switch (treeValue.value.type) {
            case "identifier": {
              const result = this.handleIdentifier(payload, treeValue.value);
              if (isUndefined(result)) {
                return;
              }
              return this.handleSubscript(result, tree.next);
            }
            case "numericLiteral": {
              const result = this.handleDotNumericLiteral(payload, treeValue.value);
              if (isUndefined(result)) {
                return;
              }
              return this.handleSubscript(result, tree.next);
            }
            case "wildcard": {
              const result = this.handleWildcard(payload);
              return this.concatIndefiniteValuePaths(result, tree.next);
            }
          }
          return;
        }
        /* c8 ignore stop */
        case "dotdot": {
          const result = this.handleDotdot(payload, treeValue);
          return this.concatIndefiniteValuePaths(result, tree.next);
        }
      }
    };
    // ============================================
    // Public Entry Point
    // ============================================
    this.handleRoot = (tree) => {
      return this.handleSubscript(this.rootPayload, tree.next);
    };
    this.rootPayload = { value: rootPayload, paths: "$" };
  }
};

// src/handler/paths.ts
var paths = (payload, path, options = {}) => {
  try {
    const tree = parseInternal(path);
    const handler = new Handler(payload);
    const result = handler.handleRoot(tree);
    if (isUndefined(result)) {
      return [];
    }
    if (!isArray(result.paths)) {
      return [result.paths];
    }
    return result.paths;
  } catch (e) {
    if (!options.hideExceptions) {
      throw e;
    }
    return [];
  }
};

// src/handler/query.ts
var querySingle = (payload, path, options) => {
  const tree = parseInternal(path);
  const handler = new Handler(payload);
  const result = handler.handleRoot(tree);
  if (!(result == null ? void 0 : result.isIndefinite) && options.returnArray) {
    if (isUndefined(result)) {
      return [];
    }
    return [result.value];
  }
  return result == null ? void 0 : result.value;
};
var queryMany = (payload, paths2, options) => {
  const results = [];
  for (const path of paths2) {
    const res = querySingle(payload, path, options);
    if (isDefined(res)) {
      results.push(res);
    }
  }
  return results;
};
var query = (payload, paths2, options = {}) => {
  try {
    if (isArray(paths2)) {
      return queryMany(payload, paths2, options);
    }
    return querySingle(payload, paths2, options);
  } catch (e) {
    if (!options.hideExceptions) {
      throw e;
    }
    return options.returnArray ? [] : void 0;
  }
};

// src/parser/stringify.ts
var COMPARATOR_OPERATORS = {
  eq: "==",
  ne: "!=",
  lt: "<",
  le: "<=",
  gt: ">",
  ge: ">=",
  in: "in",
  reg: "=~",
  nin: "nin",
  subsetof: "subsetof",
  anyof: "anyof",
  noneof: "noneof",
  empty: "empty",
  sizeof: "sizeof",
  size: "size"
};
var ARITHMETIC_OPERATORS = {
  plus: "+",
  minus: "-",
  multi: "*",
  div: "/"
};
var LOGICAL_OPERATORS = {
  and: "&&",
  or: "||"
};
var escapeForSingleQuote = (str) => {
  let result = "";
  for (let i = 0; i < str.length; i++) {
    const char = str[i];
    const code = char.charCodeAt(0);
    if (char === "\\") {
      result += "\\\\";
    } else if (char === "'") {
      result += "\\'";
    } else if (code < 32) {
      switch (char) {
        case "\b":
          result += "\\b";
          break;
        case "\f":
          result += "\\f";
          break;
        case "\n":
          result += "\\n";
          break;
        case "\r":
          result += "\\r";
          break;
        case "	":
          result += "\\t";
          break;
        default:
          result += "\\u" + code.toString(16).padStart(4, "0");
      }
    } else {
      result += char;
    }
  }
  return result;
};
var singleQuoted = (str) => `'${escapeForSingleQuote(str)}'`;
var stringifyJsonValue = (value2) => {
  if (typeof value2 === "string") {
    return singleQuoted(value2);
  }
  if (Array.isArray(value2)) {
    return `[${value2.map(stringifyJsonValue).join(",")}]`;
  }
  if (value2 !== null && typeof value2 === "object") {
    const pairs = Object.entries(value2).map(([k, v]) => `${singleQuoted(k)}:${stringifyJsonValue(v)}`);
    return `{${pairs.join(",")}}`;
  }
  return JSON.stringify(value2);
};
function stringifyValue(input) {
  switch (input.subtype) {
    case "regex":
      return `${input.value}${input.opts}`;
    case "string":
      return singleQuoted(input.value);
    case "object":
      return stringifyJsonValue(input.value);
    case "array":
      return stringifyJsonValue(input.value);
    case "boolean":
    case "number":
    case "null":
      return JSON.stringify(input.value);
    /* c8 ignore start */
    default: {
      const _exhaustive = input;
      throw new Error(`Unknown value subtype: ${_exhaustive.subtype}`);
    }
  }
}
function stringifySlice(input) {
  const start = input.start !== null ? String(input.start) : "";
  const end = input.end !== null ? String(input.end) : "";
  const step = input.step !== null ? ":" + input.step : "";
  return `${start}:${end}${step}`;
}
function stringify(input) {
  if (input === null) {
    return "";
  }
  switch (input.type) {
    case "root":
      return "$" + stringify(input.next);
    case "current":
      return "@" + stringify(input.next);
    case "dot":
      return "." + stringify(input.value);
    case "dotdot":
      return ".." + stringify(input.value);
    case "bracketExpression":
    case "bracketMember":
      return "[" + stringify(input.value) + "]";
    case "subscript":
      return stringify(input.value) + stringify(input.next);
    case "identifier":
      return input.value;
    case "wildcard":
      return "*";
    case "unions":
      return input.values.map(stringify).join(", ");
    case "stringLiteral":
      return singleQuoted(input.value);
    case "numericLiteral":
      return String(input.value);
    case "notExpression":
      return "!(" + stringify(input.value) + ")";
    case "value":
      return stringifyValue(input);
    case "filterExpression":
      return "?(" + stringify(input.value) + ")";
    case "groupOperation":
    case "groupExpression":
      return "(" + stringify(input.value) + ")";
    case "operation":
      return stringify(input.left) + " " + ARITHMETIC_OPERATORS[input.operator] + " " + stringify(input.right);
    case "comparator":
      if (input.operator === "empty") {
        return stringify(input.left) + " empty";
      }
      return stringify(input.left) + ` ${COMPARATOR_OPERATORS[input.operator]} ` + stringify(input.right);
    case "logicalExpression":
      return stringify(input.left) + ` ${LOGICAL_OPERATORS[input.operator]} ` + stringify(input.right);
    case "slices":
      return stringifySlice(input);
    case "functionCall": {
      const args = input.args.map((arg) => stringify(arg)).join(", ");
      return `${input.name}(${args})`;
    }
    /* c8 ignore start */
    default: {
      const _exhaustive = input;
      throw new Error(`Unknown AST node type: ${_exhaustive.type}`);
    }
  }
}
